{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas, copy\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "from src.utils import build_exact_table, build_bootstrapped_table, create_predictions_table, plot_truthtables\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "who_drugs = list(pandas.read_csv('dat/drugs/who2_drugs.csv').drug)\n",
    "plate_drugs = list(pandas.read_csv('dat/drugs/plate_drugs.csv').drug)\n",
    "other_drugs = list(pandas.read_csv('dat/drugs/other_drugs.csv').drug)\n",
    "mgit_drugs = list(pandas.read_csv('dat/drugs/mgit_drugs.csv').drug)\n",
    "\n",
    "drug_names_table = pandas.read_csv(\"dat/drugs/drug_names_lookup.csv\")\n",
    "drug_names_table.set_index(\"DRUG\", inplace=True)\n",
    "drug_names_lookup = {}\n",
    "for idx, row in drug_names_table.iterrows():\n",
    "    drug_names_lookup[idx] = row.DRUG_NAME.capitalize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>PHENOTYPE_METHOD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            UNIQUEID  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                 \n",
       "ERR13286038       BDQ   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "                  LZD   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "\n",
       "                       BINARY_PHENOTYPE PHENOTYPE_QUALITY PHENOTYPE_METHOD  \n",
       "ENA_RUN_ACCESSION DRUG                                                      \n",
       "ERR13286038       BDQ                 R               NaN             MGIT  \n",
       "                  LZD                 S               NaN             MGIT  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES = pandas.read_csv('dat/PHENOTYPES.csv')\n",
    "PHENOTYPES.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "PHENOTYPES.sort_index(inplace=True)\n",
    "PHENOTYPES[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR4829376</th>\n",
       "      <th>AMI</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       PREDICTION\n",
       "ENA_RUN_ACCESSION DRUG           \n",
       "ERR4829376        AMI           S\n",
       "                  BDQ           S\n",
       "                  CAP           S"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_PREDICTIONS = pandas.read_csv('dat/RAW_PREDICTIONS.csv')\n",
    "RAW_PREDICTIONS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "RAW_PREDICTIONS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 375 null calls and 1057 minor allele calls in the main effects table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>Rv0678</td>\n",
       "      <td>138_ins_g:36</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFZ</th>\n",
       "      <td>Rv0678</td>\n",
       "      <td>138_ins_g:36</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          GENE      MUTATION PREDICTION  EPISTASIS  \\\n",
       "ENA_RUN_ACCESSION DRUG                                               \n",
       "ERR13286038       BDQ   Rv0678  138_ins_g:36          S      False   \n",
       "                  CAP     tlyA          L11L          S      False   \n",
       "                  CFZ   Rv0678  138_ins_g:36          S      False   \n",
       "\n",
       "                        IS_MINOR_ALLELE  IS_NULL  \n",
       "ENA_RUN_ACCESSION DRUG                            \n",
       "ERR13286038       BDQ              True    False  \n",
       "                  CAP             False    False  \n",
       "                  CFZ              True    False  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS = pandas.read_csv('dat/RAW_EFFECTS.csv')\n",
    "RAW_EFFECTS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "RAW_EFFECTS.sort_index(inplace=True)\n",
    "print(f\"There are {RAW_EFFECTS.IS_NULL.sum()} null calls and {RAW_EFFECTS.IS_MINOR_ALLELE.sum()} minor allele calls in the main effects table\")\n",
    "RAW_EFFECTS[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most of the null calls are in the *rrs* gene so will affect the aminoglycosides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "rrs       328\n",
       "rpoB       19\n",
       "pncA        8\n",
       "Rv0678      4\n",
       "embB        4\n",
       "gid         2\n",
       "fabG1       2\n",
       "gyrB        2\n",
       "tlyA        2\n",
       "rplC        2\n",
       "ethA        1\n",
       "katG        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS[RAW_EFFECTS.IS_NULL].GENE.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whilst there are over 50 minor allele calls in genes that could affect resistance prediction for the fluorouinolones, rifampicin, bedaquline, pyrazinamide and ethambutol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "gyrA      244\n",
       "rrs       184\n",
       "Rv0678    160\n",
       "rpoB      102\n",
       "pncA       74\n",
       "fabG1      54\n",
       "embB       51\n",
       "gid        47\n",
       "gyrB       36\n",
       "katG       25\n",
       "eis        20\n",
       "rpsL       17\n",
       "ethA       14\n",
       "rplC       13\n",
       "rrl         9\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS[RAW_EFFECTS.IS_MINOR_ALLELE].GENE.value_counts()[:15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now build three EFFECTS tables that progressively include these nuances and from these build three PREDICTIONS tables that we can analyse by joining to PHENOTYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 51966/51966 [00:01<00:00, 26482.93it/s]\n",
      "100%|██████████| 52341/52341 [00:01<00:00, 26663.90it/s]\n",
      "100%|██████████| 53398/53398 [00:02<00:00, 23517.03it/s]\n"
     ]
    }
   ],
   "source": [
    "EFFECTS={}\n",
    "EFFECTS[1] = copy.deepcopy(RAW_EFFECTS[~(RAW_EFFECTS.IS_MINOR_ALLELE) & ~(RAW_EFFECTS.IS_NULL)])\n",
    "EFFECTS[1]['SET'] = 'basic'\n",
    "EFFECTS[2] = copy.deepcopy(RAW_EFFECTS[~(RAW_EFFECTS.IS_MINOR_ALLELE)])\n",
    "EFFECTS[2]['SET'] = 'nulls'\n",
    "EFFECTS[3] = copy.deepcopy(RAW_EFFECTS)\n",
    "EFFECTS[3]['SET'] = 'nulls+minors'\n",
    "assert len(EFFECTS[3]) >  len(EFFECTS[2]) >  len(EFFECTS[1])\n",
    "\n",
    "PREDICTIONS={}\n",
    "for i in [1,2,3]:\n",
    "    PREDICTIONS[i] = pandas.DataFrame(create_predictions_table(EFFECTS[i], who_drugs))\n",
    "\n",
    "assert len(RAW_PREDICTIONS) == len(PREDICTIONS[1]) == len(PREDICTIONS[2]) == len(PREDICTIONS[3])\n",
    "\n",
    "EFFECTS = pandas.concat(EFFECTS.values())\n",
    "EFFECTS.reset_index(inplace=True)\n",
    "EFFECTS.set_index(['SET','ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "EFFECTS.to_csv('dat/EFFECTS.csv')\n",
    "\n",
    "PREDICTIONS= pandas.concat(PREDICTIONS.values())\n",
    "PREDICTIONS.to_csv('dat/PREDICTIONS.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These tables each contain three scenarios (i) `basic` which ignores all null and minor alleles and is therefore closest to the WHOv2 catalogue as written, (ii) `nulls` which makes an `F` call at genetic loci associated with resistance where there are insufficient reads (two or fewer) and (iii) `nulls+minors` which adds in calls where there are three or more reads supporting a resistance-associated variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>CAP</th>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>fgd1</td>\n",
       "      <td>K270M</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              GENE MUTATION PREDICTION  EPISTASIS  \\\n",
       "SET   ENA_RUN_ACCESSION DRUG                                        \n",
       "basic ERR13286038       CAP   tlyA     L11L          S      False   \n",
       "                        DLM   fgd1    K270M          U      False   \n",
       "\n",
       "                              IS_MINOR_ALLELE  IS_NULL  \n",
       "SET   ENA_RUN_ACCESSION DRUG                            \n",
       "basic ERR13286038       CAP             False    False  \n",
       "                        DLM             False    False  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFFECTS[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             PREDICTION\n",
       "SET   ENA_RUN_ACCESSION DRUG           \n",
       "basic ERR13286038       INH           R\n",
       "                        RIF           R"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PREDICTIONS[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check that has all worked, first by picking a sample and looking at its rows in an EFFECTS table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>SET</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"22\" valign=\"top\">ERR13286038</th>\n",
       "      <th>CAP</th>\n",
       "      <td>nulls</td>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fgd1</td>\n",
       "      <td>K270M</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>nulls</td>\n",
       "      <td>embB</td>\n",
       "      <td>A409P</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>nulls</td>\n",
       "      <td>embB</td>\n",
       "      <td>E405D</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>ethA</td>\n",
       "      <td>Y211S</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>c-15t</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>c-15t</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>nulls</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H71Y</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>D103D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>H445Y</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STM</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gid</td>\n",
       "      <td>A134E</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SET   GENE MUTATION PREDICTION  EPISTASIS  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                \n",
       "ERR13286038       CAP   nulls   tlyA     L11L          S      False   \n",
       "                  DLM   nulls   fgd1    K270M          U      False   \n",
       "                  EMB   nulls   embB    A409P          U      False   \n",
       "                  EMB   nulls   embB    E405D          U      False   \n",
       "                  ETH   nulls   ethA    Y211S          U      False   \n",
       "                  ETH   nulls  fabG1    c-15t          R      False   \n",
       "                  INH   nulls  fabG1    c-15t          R      False   \n",
       "                  INH   nulls   katG    S315T          R      False   \n",
       "                  LEV   nulls   gyrA     A90V          R      False   \n",
       "                  LEV   nulls   gyrA     D94G          R      False   \n",
       "                  LEV   nulls   gyrA     E21Q          S      False   \n",
       "                  LEV   nulls   gyrA    G668D          S      False   \n",
       "                  LEV   nulls   gyrA     S95T          S      False   \n",
       "                  MXF   nulls   gyrA     A90V          R      False   \n",
       "                  MXF   nulls   gyrA     D94G          R      False   \n",
       "                  MXF   nulls   gyrA     E21Q          S      False   \n",
       "                  MXF   nulls   gyrA    G668D          S      False   \n",
       "                  MXF   nulls   gyrA     S95T          S      False   \n",
       "                  PZA   nulls   pncA     H71Y          R      False   \n",
       "                  RIF   nulls   rpoB    D103D          S      False   \n",
       "                  RIF   nulls   rpoB    H445Y          R      False   \n",
       "                  STM   nulls    gid    A134E          R      False   \n",
       "\n",
       "                        IS_MINOR_ALLELE  IS_NULL  \n",
       "ENA_RUN_ACCESSION DRUG                            \n",
       "ERR13286038       CAP             False    False  \n",
       "                  DLM             False    False  \n",
       "                  EMB             False    False  \n",
       "                  EMB             False    False  \n",
       "                  ETH             False    False  \n",
       "                  ETH             False    False  \n",
       "                  INH             False    False  \n",
       "                  INH             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  PZA             False    False  \n",
       "                  RIF             False    False  \n",
       "                  RIF             False    False  \n",
       "                  STM             False    False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_sample = 'ERR13286038'\n",
    "effects = copy.deepcopy(EFFECTS)\n",
    "effects.reset_index(inplace=True)\n",
    "effects = effects[(effects.SET=='nulls') & (effects.ENA_RUN_ACCESSION==check_sample)]\n",
    "effects.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now compare that to the calculated per drug predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/22 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:00<00:00, 21096.18it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"15\" valign=\"top\">nulls</th>\n",
       "      <th rowspan=\"15\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFZ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMI</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STM</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KAN</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             PREDICTION\n",
       "SET   ENA_RUN_ACCESSION DRUG           \n",
       "nulls ERR13286038       INH           R\n",
       "                        RIF           R\n",
       "                        PZA           R\n",
       "                        EMB           U\n",
       "                        BDQ           S\n",
       "                        LZD           S\n",
       "                        MXF           R\n",
       "                        LEV           R\n",
       "                        CFZ           S\n",
       "                        DLM           U\n",
       "                        AMI           S\n",
       "                        STM           R\n",
       "                        ETH           R\n",
       "                        KAN           S\n",
       "                        CAP           S"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_predictions_table(effects, who_drugs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we have a `RAW_PREDICTIONS` table parsed directly from the individual sample `json` files we can compare `PREDICTIONS['nulls+minors']` to that and they should be identical.\n",
    "\n",
    "A direct way of doing that is to make an antibiogram for each sample (which will automatically be in alphabetical order of drug), and then look for samples where they are different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_antibiogram(row):\n",
    "    antibiogram=''\n",
    "    for i in who_drugs:\n",
    "        antibiogram += row[('PREDICTION',i)]\n",
    "    return antibiogram\n",
    "\n",
    "df1 = RAW_PREDICTIONS.unstack()\n",
    "df1['ANTIBIOGRAM'] = df1.apply(make_antibiogram, axis=1)\n",
    "\n",
    "PREDICTIONS.reset_index(inplace=True)\n",
    "df2 = PREDICTIONS[PREDICTIONS.SET=='nulls+minors']\n",
    "PREDICTIONS.set_index(['SET','ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "df2.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "df2=df2[['PREDICTION']].unstack()\n",
    "# df2=df2.droplevel(0, axis=1)\n",
    "df2['ANTIBIOGRAM'] = df2.apply(make_antibiogram, axis=1)\n",
    "\n",
    "mask = (df1.ANTIBIOGRAM != df2.ANTIBIOGRAM)\n",
    "assert len(df1[mask]) == 0, 'some samples have different antibiograms!'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having the (alphabetical) antibiogram is useful, so let's calculate it for all three scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SET1</th>\n",
       "      <th>ANTIBIOGRAM1</th>\n",
       "      <th>SET2</th>\n",
       "      <th>ANTIBIOGRAM2</th>\n",
       "      <th>SET3</th>\n",
       "      <th>ANTIBIOGRAM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR13286038</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286039</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSRSSSRSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286042</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SET1     ANTIBIOGRAM1   SET2     ANTIBIOGRAM2  \\\n",
       "ENA_RUN_ACCESSION                                                   \n",
       "ERR13286038        basic  RRRUSSRRSUSRRSS  nulls  RRRUSSRRSUSRRSS   \n",
       "ERR13286039        basic  RSSSSSSSSSSSSSS  nulls  RSSSSSSSSSSSSSS   \n",
       "ERR13286042        basic  RRRRSURRSSRRRRR  nulls  RRRRSURRSSRRRRR   \n",
       "\n",
       "                           SET3     ANTIBIOGRAM3  \n",
       "ENA_RUN_ACCESSION                                 \n",
       "ERR13286038        nulls+minors  RRRUSSRRSUSRRSS  \n",
       "ERR13286039        nulls+minors  RSSSRSSSRSSSSSS  \n",
       "ERR13286042        nulls+minors  RRRRSURRSSRRRRR  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = PREDICTIONS.unstack()\n",
    "df['ANTIBIOGRAM'] = df.apply(make_antibiogram, axis=1)\n",
    "df = df[[('ANTIBIOGRAM','')]]\n",
    "df = df.droplevel(1, axis=1)\n",
    "df.reset_index(inplace=True)\n",
    "df1 = df[df.SET=='basic']\n",
    "df2 = df[df.SET=='nulls']\n",
    "df3 = df[df.SET=='nulls+minors']\n",
    "df1.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df2.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df3.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df1 = df1.rename(columns={'SET':\"SET1\",'ANTIBIOGRAM':'ANTIBIOGRAM1'})\n",
    "df2 = df2.rename(columns={'SET':\"SET2\",'ANTIBIOGRAM':'ANTIBIOGRAM2'})\n",
    "df3 = df3.rename(columns={'SET':\"SET3\",'ANTIBIOGRAM':'ANTIBIOGRAM3'})\n",
    "ANTIBIOGRAMS = df1.join(df2).join(df3)\n",
    "ANTIBIOGRAMS.to_csv('dat/ANTIBIOGRAMS.csv')\n",
    "ANTIBIOGRAMS[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now look to see how many samples out of 2,663 change between scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calling nulls at resistance loci changes the antibiogram of 33 samples,whilst also calling minor alleles leads to 210 samples changing antibiogram\n"
     ]
    }
   ],
   "source": [
    "print(f\"Calling nulls at resistance loci changes the antibiogram of {(ANTIBIOGRAMS.ANTIBIOGRAM1!=ANTIBIOGRAMS.ANTIBIOGRAM2).sum()} samples,\\\n",
    "whilst also calling minor alleles leads to {(ANTIBIOGRAMS.ANTIBIOGRAM2!=ANTIBIOGRAMS.ANTIBIOGRAM3).sum()} samples changing antibiogram\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('UKMYC', 'INH'),\n",
       " ('UKMYC', 'RIF'),\n",
       " ('MGIT', 'PZA'),\n",
       " ('UKMYC', 'EMB'),\n",
       " ('MGIT', 'BDQ'),\n",
       " ('UKMYC', 'LZD'),\n",
       " ('UKMYC', 'MXF'),\n",
       " ('UKMYC', 'LEV'),\n",
       " ('UKMYC', 'CFZ'),\n",
       " ('UKMYC', 'DLM'),\n",
       " ('UKMYC', 'AMI'),\n",
       " ('MGIT', 'STM'),\n",
       " ('UKMYC', 'ETH'),\n",
       " ('UKMYC', 'KAN'),\n",
       " ('MGIT', 'CAP')]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_drug_order = pandas.read_csv('dat/drugs/results_drugs.csv')\n",
    "results_drug_order.set_index(['pDST method', 'drug'], inplace=True)\n",
    "list(results_drug_order.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3/3 [00:32<00:00, 10.86s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>sensitivity_sem</th>\n",
       "      <th>specificity</th>\n",
       "      <th>specificity_sem</th>\n",
       "      <th>PPV</th>\n",
       "      <th>PPV_sem</th>\n",
       "      <th>RR</th>\n",
       "      <th>SR</th>\n",
       "      <th>UR</th>\n",
       "      <th>RS</th>\n",
       "      <th>SS</th>\n",
       "      <th>US</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set</th>\n",
       "      <th>drug</th>\n",
       "      <th>method</th>\n",
       "      <th>dataset</th>\n",
       "      <th>quality</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">INH</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">UKMYC</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>91.240875</td>\n",
       "      <td>0.480167</td>\n",
       "      <td>95.395150</td>\n",
       "      <td>0.367757</td>\n",
       "      <td>95.206597</td>\n",
       "      <td>0.370676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>93.746796</td>\n",
       "      <td>0.336085</td>\n",
       "      <td>96.624866</td>\n",
       "      <td>0.352979</td>\n",
       "      <td>96.618858</td>\n",
       "      <td>0.342746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>91.200000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.600000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.397490</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>456.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>448.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>93.913043</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.412556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.428571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>432.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>906.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <th>UKMYC</th>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>93.729501</td>\n",
       "      <td>0.406094</td>\n",
       "      <td>95.957331</td>\n",
       "      <td>0.314197</td>\n",
       "      <td>95.445435</td>\n",
       "      <td>0.375315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">nulls+minors</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">KAN</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">UKMYC</th>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>HIGH</th>\n",
       "      <td>84.298262</td>\n",
       "      <td>0.713334</td>\n",
       "      <td>98.322163</td>\n",
       "      <td>0.199675</td>\n",
       "      <td>95.369505</td>\n",
       "      <td>0.574964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>76.470588</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.730606</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.086957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>221.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>650.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>84.615385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.543689</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.069869</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>220.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>880.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CAP</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">MGIT</th>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>74.983621</td>\n",
       "      <td>0.941786</td>\n",
       "      <td>98.295397</td>\n",
       "      <td>0.200364</td>\n",
       "      <td>94.980289</td>\n",
       "      <td>0.593964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>75.098814</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.254799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>831.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 sensitivity  sensitivity_sem  \\\n",
       "set          drug method dataset        quality                                 \n",
       "basic        INH  UKMYC  bootstrapped50 ALL        91.240875         0.480167   \n",
       "                                        HIGH       93.746796         0.336085   \n",
       "                         entire         ALL        91.200000         0.000000   \n",
       "                                        HIGH       93.913043         0.000000   \n",
       "             RIF  UKMYC  bootstrapped50 ALL        93.729501         0.406094   \n",
       "...                                                      ...              ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH       84.298262         0.713334   \n",
       "                         entire         ALL        76.470588         0.000000   \n",
       "                                        HIGH       84.615385         0.000000   \n",
       "             CAP  MGIT   bootstrapped50 ALL        74.983621         0.941786   \n",
       "                         entire         ALL        75.098814         0.000000   \n",
       "\n",
       "                                                 specificity  specificity_sem  \\\n",
       "set          drug method dataset        quality                                 \n",
       "basic        INH  UKMYC  bootstrapped50 ALL        95.395150         0.367757   \n",
       "                                        HIGH       96.624866         0.352979   \n",
       "                         entire         ALL        95.600000         0.000000   \n",
       "                                        HIGH       96.412556         0.000000   \n",
       "             RIF  UKMYC  bootstrapped50 ALL        95.957331         0.314197   \n",
       "...                                                      ...              ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH       98.322163         0.199675   \n",
       "                         entire         ALL        98.730606         0.000000   \n",
       "                                        HIGH       98.543689         0.000000   \n",
       "             CAP  MGIT   bootstrapped50 ALL        98.295397         0.200364   \n",
       "                         entire         ALL        98.254799         0.000000   \n",
       "\n",
       "                                                       PPV   PPV_sem     RR  \\\n",
       "set          drug method dataset        quality                               \n",
       "basic        INH  UKMYC  bootstrapped50 ALL      95.206597  0.370676    NaN   \n",
       "                                        HIGH     96.618858  0.342746    NaN   \n",
       "                         entire         ALL      95.397490  0.000000  456.0   \n",
       "                                        HIGH     96.428571  0.000000  432.0   \n",
       "             RIF  UKMYC  bootstrapped50 ALL      95.445435  0.375315    NaN   \n",
       "...                                                    ...       ...    ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH     95.369505  0.574964    NaN   \n",
       "                         entire         ALL      96.086957  0.000000  221.0   \n",
       "                                        HIGH     96.069869  0.000000  220.0   \n",
       "             CAP  MGIT   bootstrapped50 ALL      94.980289  0.593964    NaN   \n",
       "                         entire         ALL      95.000000  0.000000  190.0   \n",
       "\n",
       "                                                   SR    UR    RS     SS  \\\n",
       "set          drug method dataset        quality                            \n",
       "basic        INH  UKMYC  bootstrapped50 ALL       NaN   NaN   NaN    NaN   \n",
       "                                        HIGH      NaN   NaN   NaN    NaN   \n",
       "                         entire         ALL      33.0  11.0  22.0  448.0   \n",
       "                                        HIGH     23.0   5.0  16.0  404.0   \n",
       "             RIF  UKMYC  bootstrapped50 ALL       NaN   NaN   NaN    NaN   \n",
       "...                                               ...   ...   ...    ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH      NaN   NaN   NaN    NaN   \n",
       "                         entire         ALL      56.0  12.0   9.0  650.0   \n",
       "                                        HIGH     34.0   6.0   9.0  563.0   \n",
       "             CAP  MGIT   bootstrapped50 ALL       NaN   NaN   NaN    NaN   \n",
       "                         entire         ALL      58.0   5.0  10.0  536.0   \n",
       "\n",
       "                                                   US   Total  \n",
       "set          drug method dataset        quality                \n",
       "basic        INH  UKMYC  bootstrapped50 ALL       NaN     NaN  \n",
       "                                        HIGH      NaN     NaN  \n",
       "                         entire         ALL      30.0  1000.0  \n",
       "                                        HIGH     26.0   906.0  \n",
       "             RIF  UKMYC  bootstrapped50 ALL       NaN     NaN  \n",
       "...                                               ...     ...  \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH      NaN     NaN  \n",
       "                         entire         ALL      50.0  1000.0  \n",
       "                                        HIGH     46.0   880.0  \n",
       "             CAP  MGIT   bootstrapped50 ALL       NaN     NaN  \n",
       "                         entire         ALL      27.0   831.0  \n",
       "\n",
       "[156 rows x 13 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results={}\n",
    "PREDICTIONS.reset_index(inplace=True)\n",
    "PREDICTIONS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "\n",
    "for i in tqdm(['basic','nulls','nulls+minors']):\n",
    "    \n",
    "    predictions = PREDICTIONS[PREDICTIONS.SET==i]\n",
    "    df = copy.deepcopy(predictions.join(PHENOTYPES, how='inner'))\n",
    "\n",
    "    plate_df = df[df.PHENOTYPE_METHOD=='UKMYC']\n",
    "    plate_df.reset_index(inplace=True)\n",
    "    table = []\n",
    "    table = build_exact_table(table, plate_df, plate_drugs, ['HIGH', 'ALL'], i, 'UKMYC', include_fails=False)\n",
    "    table = build_bootstrapped_table(table,plate_df, plate_drugs, ['HIGH', 'ALL'], i, 'UKMYC', include_fails=False)\n",
    "\n",
    "    mgit_df = df[df.PHENOTYPE_METHOD=='MGIT']\n",
    "    mgit_df.reset_index(inplace=True)\n",
    "    table = build_exact_table(table, mgit_df, mgit_drugs, ['ALL'], i, 'MGIT', include_fails=False)\n",
    "    table = build_bootstrapped_table(table,mgit_df, mgit_drugs, ['ALL'], i, 'MGIT', include_fails=False)\n",
    "    results[i] = pandas.DataFrame(table, columns=['set','drug','method','quality','dataset','sensitivity','sensitivity_sem','specificity','specificity_sem','PPV','PPV_sem', 'RR', 'SR', 'UR', 'RS', 'SS', 'US', 'Total'])\n",
    "\n",
    "    results[i].set_index(['method','drug'],inplace=True)\n",
    "    results[i] = results[i][results[i].index.isin(list(results_drug_order.index))]\n",
    "    results[i].reset_index(inplace=True)\n",
    "    results[i].drug = results[i].drug.astype('category')\n",
    "    results[i].drug = results[i].drug.cat.set_categories(who_drugs)\n",
    "    results[i].sort_values(['drug', 'dataset', 'quality'], inplace=True)\n",
    "    results[i].set_index(['set','drug','method','dataset', 'quality'],inplace=True)\n",
    "\n",
    "results = pandas.concat(results.values())\n",
    "results.to_csv('dat/RESULTS.csv')\n",
    "results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>method</th>\n",
       "      <th>drug</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">UKMYC</th>\n",
       "      <th>INH</th>\n",
       "      <td>92.5 ±0.5</td>\n",
       "      <td>94.6 ±0.4</td>\n",
       "      <td>94.5 ±0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>95.3 ±0.4</td>\n",
       "      <td>94.9 ±0.3</td>\n",
       "      <td>94.5 ±0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGIT</th>\n",
       "      <th>PZA</th>\n",
       "      <td>85.1 ±0.7</td>\n",
       "      <td>96.7 ±0.4</td>\n",
       "      <td>95.7 ±0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UKMYC</th>\n",
       "      <th>EMB</th>\n",
       "      <td>86.7 ±0.8</td>\n",
       "      <td>84.1 ±0.5</td>\n",
       "      <td>68.4 ±1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGIT</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>46.1 ±1.0</td>\n",
       "      <td>98.6 ±0.2</td>\n",
       "      <td>97.2 ±0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">UKMYC</th>\n",
       "      <th>LZD</th>\n",
       "      <td>28.1 ±1.9</td>\n",
       "      <td>99.8 ±0.1</td>\n",
       "      <td>94.3 ±1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>85.8 ±0.8</td>\n",
       "      <td>93.8 ±0.4</td>\n",
       "      <td>84.9 ±0.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>82.3 ±0.7</td>\n",
       "      <td>96.4 ±0.3</td>\n",
       "      <td>91.7 ±0.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFZ</th>\n",
       "      <td>8.1 ±0.6</td>\n",
       "      <td>98.0 ±0.2</td>\n",
       "      <td>62.4 ±3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>12.3 ±1.1</td>\n",
       "      <td>99.9 ±0.0</td>\n",
       "      <td>95.5 ±1.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMI</th>\n",
       "      <td>74.3 ±1.0</td>\n",
       "      <td>99.4 ±0.1</td>\n",
       "      <td>98.1 ±0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGIT</th>\n",
       "      <th>STM</th>\n",
       "      <td>80.7 ±0.8</td>\n",
       "      <td>95.3 ±0.4</td>\n",
       "      <td>92.3 ±0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">UKMYC</th>\n",
       "      <th>ETH</th>\n",
       "      <td>72.2 ±1.0</td>\n",
       "      <td>87.0 ±0.5</td>\n",
       "      <td>69.4 ±1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KAN</th>\n",
       "      <td>76.7 ±1.0</td>\n",
       "      <td>98.7 ±0.2</td>\n",
       "      <td>96.0 ±0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGIT</th>\n",
       "      <th>CAP</th>\n",
       "      <td>75.0 ±0.9</td>\n",
       "      <td>98.3 ±0.2</td>\n",
       "      <td>95.0 ±0.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            sensitivity specificity        PPV\n",
       "method drug                                   \n",
       "UKMYC  INH    92.5 ±0.5   94.6 ±0.4  94.5 ±0.4\n",
       "       RIF    95.3 ±0.4   94.9 ±0.3  94.5 ±0.4\n",
       "MGIT   PZA    85.1 ±0.7   96.7 ±0.4  95.7 ±0.5\n",
       "UKMYC  EMB    86.7 ±0.8   84.1 ±0.5  68.4 ±1.0\n",
       "MGIT   BDQ    46.1 ±1.0   98.6 ±0.2  97.2 ±0.4\n",
       "UKMYC  LZD    28.1 ±1.9   99.8 ±0.1  94.3 ±1.9\n",
       "       MXF    85.8 ±0.8   93.8 ±0.4  84.9 ±0.9\n",
       "       LEV    82.3 ±0.7   96.4 ±0.3  91.7 ±0.7\n",
       "       CFZ     8.1 ±0.6   98.0 ±0.2  62.4 ±3.1\n",
       "       DLM    12.3 ±1.1   99.9 ±0.0  95.5 ±1.8\n",
       "       AMI    74.3 ±1.0   99.4 ±0.1  98.1 ±0.3\n",
       "MGIT   STM    80.7 ±0.8   95.3 ±0.4  92.3 ±0.6\n",
       "UKMYC  ETH    72.2 ±1.0   87.0 ±0.5  69.4 ±1.0\n",
       "       KAN    76.7 ±1.0   98.7 ±0.2  96.0 ±0.6\n",
       "MGIT   CAP    75.0 ±0.9   98.3 ±0.2  95.0 ±0.6"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.reset_index(inplace=True)\n",
    "df = copy.deepcopy(results[results.set=='nulls+minors'])\n",
    "df = df[(df.dataset=='bootstrapped50') & (df.quality=='ALL')]\n",
    "df = df[['method', 'drug', 'sensitivity',\n",
    "       'sensitivity_sem', 'specificity', 'specificity_sem', 'PPV', 'PPV_sem']]\n",
    "\n",
    "for col in ['sensitivity', 'sensitivity_sem', 'specificity', 'specificity_sem', 'PPV', 'PPV_sem']:\n",
    "    df[col] = df[col].map('{:,.1f}'.format)\n",
    "for col in ['sensitivity', 'specificity', 'PPV']:\n",
    "    df[col] = df[col] + ' ±' + df[col+'_sem']\n",
    "df = df[['method', 'drug', 'sensitivity',\n",
    "        'specificity', 'PPV', ]]\n",
    "df.set_index(['method', 'drug'], inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{llrrr}\n",
      "\\toprule\n",
      " &  & sensitivity & specificity & PPV \\\\\n",
      "method & drug &  &  &  \\\\\n",
      "\\midrule\n",
      "\\multirow[t]{2}{*}{UKMYC} & INH & 92.5 ±0.5 & 94.6 ±0.4 & 94.5 ±0.4 \\\\\n",
      " & RIF & 95.3 ±0.4 & 94.9 ±0.3 & 94.5 ±0.4 \\\\\n",
      "\\cline{1-5}\n",
      "MGIT & PZA & 85.1 ±0.7 & 96.7 ±0.4 & 95.7 ±0.5 \\\\\n",
      "\\cline{1-5}\n",
      "UKMYC & EMB & 86.7 ±0.8 & 84.1 ±0.5 & 68.4 ±1.0 \\\\\n",
      "\\cline{1-5}\n",
      "MGIT & BDQ & 46.1 ±1.0 & 98.6 ±0.2 & 97.2 ±0.4 \\\\\n",
      "\\cline{1-5}\n",
      "\\multirow[t]{6}{*}{UKMYC} & LZD & 28.1 ±1.9 & 99.8 ±0.1 & 94.3 ±1.9 \\\\\n",
      " & MXF & 85.8 ±0.8 & 93.8 ±0.4 & 84.9 ±0.9 \\\\\n",
      " & LEV & 82.3 ±0.7 & 96.4 ±0.3 & 91.7 ±0.7 \\\\\n",
      " & CFZ & 8.1 ±0.6 & 98.0 ±0.2 & 62.4 ±3.1 \\\\\n",
      " & DLM & 12.3 ±1.1 & 99.9 ±0.0 & 95.5 ±1.8 \\\\\n",
      " & AMI & 74.3 ±1.0 & 99.4 ±0.1 & 98.1 ±0.3 \\\\\n",
      "\\cline{1-5}\n",
      "MGIT & STM & 80.7 ±0.8 & 95.3 ±0.4 & 92.3 ±0.6 \\\\\n",
      "\\cline{1-5}\n",
      "\\multirow[t]{2}{*}{UKMYC} & ETH & 72.2 ±1.0 & 87.0 ±0.5 & 69.4 ±1.0 \\\\\n",
      " & KAN & 76.7 ±1.0 & 98.7 ±0.2 & 96.0 ±0.6 \\\\\n",
      "\\cline{1-5}\n",
      "MGIT & CAP & 75.0 ±0.9 & 98.3 ±0.2 & 95.0 ±0.6 \\\\\n",
      "\\cline{1-5}\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(df.to_latex(column_format='llrrr', multirow=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "for set in ['basic', 'nulls', 'nulls+minors']:\n",
    "\n",
    "    df = results[(results.set==set)]\n",
    "\n",
    "    plot_truthtables(df, ['ALL','HIGH'], f'table-{set}-')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>PHENOTYPE_METHOD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286039</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>site.10.subj.JE01389322.lab.JE01389322.iso.1</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            UNIQUEID  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                 \n",
       "ERR13286038       BDQ   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "                  LZD   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "ERR13286039       BDQ   site.10.subj.JE01389322.lab.JE01389322.iso.1   \n",
       "\n",
       "                       BINARY_PHENOTYPE PHENOTYPE_QUALITY PHENOTYPE_METHOD  \n",
       "ENA_RUN_ACCESSION DRUG                                                      \n",
       "ERR13286038       BDQ                 R               NaN             MGIT  \n",
       "                  LZD                 S               NaN             MGIT  \n",
       "ERR13286039       BDQ                 R               NaN             MGIT  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>SET</th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>basic</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>basic</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>basic</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SET PREDICTION\n",
       "ENA_RUN_ACCESSION DRUG                  \n",
       "ERR13286038       INH   basic          R\n",
       "                  RIF   basic          R\n",
       "                  PZA   basic          R"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PREDICTIONS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>ENA_STUDY_ACCESSION</th>\n",
       "      <th>ENA_SAMPLE_ACCESSION</th>\n",
       "      <th>HAS_ALL_DRUGS</th>\n",
       "      <th>N_SUSCEPTIBLE</th>\n",
       "      <th>N_RESISTANT</th>\n",
       "      <th>N_HIGH_QUALITY</th>\n",
       "      <th>AMI</th>\n",
       "      <th>BDQ</th>\n",
       "      <th>...</th>\n",
       "      <th>EMB_QUALITY</th>\n",
       "      <th>ETH_QUALITY</th>\n",
       "      <th>INH_QUALITY</th>\n",
       "      <th>KAN_QUALITY</th>\n",
       "      <th>LEV_QUALITY</th>\n",
       "      <th>LZD_QUALITY</th>\n",
       "      <th>MXF_QUALITY</th>\n",
       "      <th>RIF_QUALITY</th>\n",
       "      <th>FASTQ_FTP</th>\n",
       "      <th>POS_AVG_GROWTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR4810791</td>\n",
       "      <td>site.02.subj.0068.lab.22A018.iso.1</td>\n",
       "      <td>PRJEB41194</td>\n",
       "      <td>SAMEA7542366</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR481/001/ERR481...</td>\n",
       "      <td>44.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR4810943</td>\n",
       "      <td>site.02.subj.0091.lab.22A043.iso.1</td>\n",
       "      <td>PRJEB41194</td>\n",
       "      <td>SAMEA7544003</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LOW</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR481/003/ERR481...</td>\n",
       "      <td>43.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR4810948</td>\n",
       "      <td>site.02.subj.0096.lab.22A048.iso.1</td>\n",
       "      <td>PRJEB41194</td>\n",
       "      <td>SAMEA7544008</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR481/008/ERR481...</td>\n",
       "      <td>26.20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ENA_RUN_ACCESSION                            UNIQUEID ENA_STUDY_ACCESSION  \\\n",
       "0        ERR4810791  site.02.subj.0068.lab.22A018.iso.1          PRJEB41194   \n",
       "1        ERR4810943  site.02.subj.0091.lab.22A043.iso.1          PRJEB41194   \n",
       "2        ERR4810948  site.02.subj.0096.lab.22A048.iso.1          PRJEB41194   \n",
       "\n",
       "  ENA_SAMPLE_ACCESSION  HAS_ALL_DRUGS  N_SUSCEPTIBLE  N_RESISTANT  \\\n",
       "0         SAMEA7542366           True              6            6   \n",
       "1         SAMEA7544003           True              5            7   \n",
       "2         SAMEA7544008           True              5            7   \n",
       "\n",
       "   N_HIGH_QUALITY AMI BDQ  ... EMB_QUALITY ETH_QUALITY INH_QUALITY  \\\n",
       "0               9   S   S  ...        HIGH        HIGH        HIGH   \n",
       "1              11   S   S  ...        HIGH        HIGH        HIGH   \n",
       "2              11   S   R  ...        HIGH        HIGH        HIGH   \n",
       "\n",
       "  KAN_QUALITY LEV_QUALITY LZD_QUALITY MXF_QUALITY RIF_QUALITY  \\\n",
       "0        HIGH         LOW         LOW        HIGH        HIGH   \n",
       "1        HIGH        HIGH         LOW        HIGH        HIGH   \n",
       "2        HIGH        HIGH        HIGH        HIGH        HIGH   \n",
       "\n",
       "                                           FASTQ_FTP POS_AVG_GROWTH  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/ERR481/001/ERR481...          44.68  \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR481/003/ERR481...          43.23  \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/ERR481/008/ERR481...          26.20  \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UKMYC = pandas.read_csv('dat/UKMYC_1000_samples.csv')\n",
    "UKMYC[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
