{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, copy\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "from src.utils import build_exact_table, build_bootstrapped_table, create_predictions_table, plot_truthtables, plot_growth_boxplot\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "who_drugs = list(pandas.read_csv('dat/drugs/who2_drugs.csv').drug)\n",
    "plate_drugs = list(pandas.read_csv('dat/drugs/plate_drugs.csv').drug)\n",
    "other_drugs = list(pandas.read_csv('dat/drugs/other_drugs.csv').drug)\n",
    "mgit_drugs = list(pandas.read_csv('dat/drugs/mgit_drugs.csv').drug)\n",
    "\n",
    "drug_names_table = pandas.read_csv(\"dat/drugs/drug_names_lookup.csv\")\n",
    "drug_names_table.set_index(\"DRUG\", inplace=True)\n",
    "drug_names_lookup = {}\n",
    "for idx, row in drug_names_table.iterrows():\n",
    "    drug_names_lookup[idx] = row.DRUG_NAME.capitalize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>PHENOTYPE_METHOD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            UNIQUEID  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                 \n",
       "ERR13286038       BDQ   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "                  LZD   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "\n",
       "                       BINARY_PHENOTYPE PHENOTYPE_QUALITY PHENOTYPE_METHOD  \n",
       "ENA_RUN_ACCESSION DRUG                                                      \n",
       "ERR13286038       BDQ                 R               NaN             MGIT  \n",
       "                  LZD                 S               NaN             MGIT  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES = pandas.read_csv('dat/PHENOTYPES.csv')\n",
    "PHENOTYPES.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "PHENOTYPES.sort_index(inplace=True)\n",
    "PHENOTYPES[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR4829376</th>\n",
       "      <th>AMI</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       PREDICTION\n",
       "ENA_RUN_ACCESSION DRUG           \n",
       "ERR4829376        AMI           S\n",
       "                  BDQ           S\n",
       "                  CAP           S"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_PREDICTIONS = pandas.read_csv('dat/RAW_PREDICTIONS.csv')\n",
    "RAW_PREDICTIONS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "RAW_PREDICTIONS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 375 null calls and 1057 minor allele calls in the main effects table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>Rv0678</td>\n",
       "      <td>138_ins_g:36</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFZ</th>\n",
       "      <td>Rv0678</td>\n",
       "      <td>138_ins_g:36</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          GENE      MUTATION PREDICTION  EPISTASIS  \\\n",
       "ENA_RUN_ACCESSION DRUG                                               \n",
       "ERR13286038       BDQ   Rv0678  138_ins_g:36          S      False   \n",
       "                  CAP     tlyA          L11L          S      False   \n",
       "                  CFZ   Rv0678  138_ins_g:36          S      False   \n",
       "\n",
       "                        IS_MINOR_ALLELE  IS_NULL  \n",
       "ENA_RUN_ACCESSION DRUG                            \n",
       "ERR13286038       BDQ              True    False  \n",
       "                  CAP             False    False  \n",
       "                  CFZ              True    False  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS = pandas.read_csv('dat/RAW_EFFECTS.csv')\n",
    "RAW_EFFECTS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "RAW_EFFECTS.sort_index(inplace=True)\n",
    "print(f\"There are {RAW_EFFECTS.IS_NULL.sum()} null calls and {RAW_EFFECTS.IS_MINOR_ALLELE.sum()} minor allele calls in the main effects table\")\n",
    "RAW_EFFECTS[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most of the null calls are in the *rrs* gene so will affect the aminoglycosides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "rrs       328\n",
       "rpoB       19\n",
       "pncA        8\n",
       "Rv0678      4\n",
       "embB        4\n",
       "gid         2\n",
       "fabG1       2\n",
       "gyrB        2\n",
       "tlyA        2\n",
       "rplC        2\n",
       "ethA        1\n",
       "katG        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS[RAW_EFFECTS.IS_NULL].GENE.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whilst there are over 50 minor allele calls in genes that could affect resistance prediction for the fluorouinolones, rifampicin, bedaquline, pyrazinamide and ethambutol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "gyrA      244\n",
       "rrs       184\n",
       "Rv0678    160\n",
       "rpoB      102\n",
       "pncA       74\n",
       "fabG1      54\n",
       "embB       51\n",
       "gid        47\n",
       "gyrB       36\n",
       "katG       25\n",
       "eis        20\n",
       "rpsL       17\n",
       "ethA       14\n",
       "rplC       13\n",
       "rrl         9\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS[RAW_EFFECTS.IS_MINOR_ALLELE].GENE.value_counts()[:15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now build three EFFECTS tables that progressively include these nuances and from these build three PREDICTIONS tables that we can analyse by joining to PHENOTYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 51966/51966 [00:01<00:00, 26549.73it/s]\n",
      "100%|██████████| 52341/52341 [00:01<00:00, 26611.48it/s]\n",
      "100%|██████████| 53398/53398 [00:02<00:00, 26604.38it/s]\n"
     ]
    }
   ],
   "source": [
    "EFFECTS={}\n",
    "EFFECTS[1] = copy.deepcopy(RAW_EFFECTS[~(RAW_EFFECTS.IS_MINOR_ALLELE) & ~(RAW_EFFECTS.IS_NULL)])\n",
    "EFFECTS[1]['SET'] = 'basic'\n",
    "EFFECTS[2] = copy.deepcopy(RAW_EFFECTS[~(RAW_EFFECTS.IS_MINOR_ALLELE)])\n",
    "EFFECTS[2]['SET'] = 'nulls'\n",
    "EFFECTS[3] = copy.deepcopy(RAW_EFFECTS)\n",
    "EFFECTS[3]['SET'] = 'nulls+minors'\n",
    "assert len(EFFECTS[3]) >  len(EFFECTS[2]) >  len(EFFECTS[1])\n",
    "\n",
    "PREDICTIONS={}\n",
    "for i in [1,2,3]:\n",
    "    PREDICTIONS[i] = pandas.DataFrame(create_predictions_table(EFFECTS[i], who_drugs))\n",
    "\n",
    "assert len(RAW_PREDICTIONS) == len(PREDICTIONS[1]) == len(PREDICTIONS[2]) == len(PREDICTIONS[3])\n",
    "\n",
    "EFFECTS = pandas.concat(EFFECTS.values())\n",
    "EFFECTS.reset_index(inplace=True)\n",
    "EFFECTS.set_index(['SET','ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "EFFECTS.to_csv('dat/EFFECTS.csv')\n",
    "\n",
    "PREDICTIONS= pandas.concat(PREDICTIONS.values())\n",
    "PREDICTIONS.to_csv('dat/PREDICTIONS.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These tables each contain three scenarios (i) `basic` which ignores all null and minor alleles and is therefore closest to the WHOv2 catalogue as written, (ii) `nulls` which makes an `F` call at genetic loci associated with resistance where there are insufficient reads (two or fewer) and (iii) `nulls+minors` which adds in calls where there are three or more reads supporting a resistance-associated variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>CAP</th>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>fgd1</td>\n",
       "      <td>K270M</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              GENE MUTATION PREDICTION  EPISTASIS  \\\n",
       "SET   ENA_RUN_ACCESSION DRUG                                        \n",
       "basic ERR13286038       CAP   tlyA     L11L          S      False   \n",
       "                        DLM   fgd1    K270M          U      False   \n",
       "\n",
       "                              IS_MINOR_ALLELE  IS_NULL  \n",
       "SET   ENA_RUN_ACCESSION DRUG                            \n",
       "basic ERR13286038       CAP             False    False  \n",
       "                        DLM             False    False  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFFECTS[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             PREDICTION\n",
       "SET   ENA_RUN_ACCESSION DRUG           \n",
       "basic ERR13286038       INH           R\n",
       "                        RIF           R"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PREDICTIONS[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check that has all worked, first by picking a sample and looking at its rows in an EFFECTS table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>SET</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"22\" valign=\"top\">ERR13286038</th>\n",
       "      <th>CAP</th>\n",
       "      <td>nulls</td>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fgd1</td>\n",
       "      <td>K270M</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>nulls</td>\n",
       "      <td>embB</td>\n",
       "      <td>A409P</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>nulls</td>\n",
       "      <td>embB</td>\n",
       "      <td>E405D</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>ethA</td>\n",
       "      <td>Y211S</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>c-15t</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>c-15t</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>nulls</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H71Y</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>D103D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>H445Y</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STM</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gid</td>\n",
       "      <td>A134E</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SET   GENE MUTATION PREDICTION  EPISTASIS  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                \n",
       "ERR13286038       CAP   nulls   tlyA     L11L          S      False   \n",
       "                  DLM   nulls   fgd1    K270M          U      False   \n",
       "                  EMB   nulls   embB    A409P          U      False   \n",
       "                  EMB   nulls   embB    E405D          U      False   \n",
       "                  ETH   nulls   ethA    Y211S          U      False   \n",
       "                  ETH   nulls  fabG1    c-15t          R      False   \n",
       "                  INH   nulls  fabG1    c-15t          R      False   \n",
       "                  INH   nulls   katG    S315T          R      False   \n",
       "                  LEV   nulls   gyrA     A90V          R      False   \n",
       "                  LEV   nulls   gyrA     D94G          R      False   \n",
       "                  LEV   nulls   gyrA     E21Q          S      False   \n",
       "                  LEV   nulls   gyrA    G668D          S      False   \n",
       "                  LEV   nulls   gyrA     S95T          S      False   \n",
       "                  MXF   nulls   gyrA     A90V          R      False   \n",
       "                  MXF   nulls   gyrA     D94G          R      False   \n",
       "                  MXF   nulls   gyrA     E21Q          S      False   \n",
       "                  MXF   nulls   gyrA    G668D          S      False   \n",
       "                  MXF   nulls   gyrA     S95T          S      False   \n",
       "                  PZA   nulls   pncA     H71Y          R      False   \n",
       "                  RIF   nulls   rpoB    D103D          S      False   \n",
       "                  RIF   nulls   rpoB    H445Y          R      False   \n",
       "                  STM   nulls    gid    A134E          R      False   \n",
       "\n",
       "                        IS_MINOR_ALLELE  IS_NULL  \n",
       "ENA_RUN_ACCESSION DRUG                            \n",
       "ERR13286038       CAP             False    False  \n",
       "                  DLM             False    False  \n",
       "                  EMB             False    False  \n",
       "                  EMB             False    False  \n",
       "                  ETH             False    False  \n",
       "                  ETH             False    False  \n",
       "                  INH             False    False  \n",
       "                  INH             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  PZA             False    False  \n",
       "                  RIF             False    False  \n",
       "                  RIF             False    False  \n",
       "                  STM             False    False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_sample = 'ERR13286038'\n",
    "effects = copy.deepcopy(EFFECTS)\n",
    "effects.reset_index(inplace=True)\n",
    "effects = effects[(effects.SET=='nulls') & (effects.ENA_RUN_ACCESSION==check_sample)]\n",
    "effects.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now compare that to the calculated per drug predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/22 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:00<00:00, 24712.02it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"15\" valign=\"top\">nulls</th>\n",
       "      <th rowspan=\"15\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFZ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMI</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STM</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KAN</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             PREDICTION\n",
       "SET   ENA_RUN_ACCESSION DRUG           \n",
       "nulls ERR13286038       INH           R\n",
       "                        RIF           R\n",
       "                        PZA           R\n",
       "                        EMB           U\n",
       "                        BDQ           S\n",
       "                        LZD           S\n",
       "                        MXF           R\n",
       "                        LEV           R\n",
       "                        CFZ           S\n",
       "                        DLM           U\n",
       "                        AMI           S\n",
       "                        STM           R\n",
       "                        ETH           R\n",
       "                        KAN           S\n",
       "                        CAP           S"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_predictions_table(effects, who_drugs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we have a `RAW_PREDICTIONS` table parsed directly from the individual sample `json` files we can compare `PREDICTIONS['nulls+minors']` to that and they should be identical.\n",
    "\n",
    "A direct way of doing that is to make an antibiogram for each sample (which will automatically be in alphabetical order of drug), and then look for samples where they are different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_antibiogram(row):\n",
    "    antibiogram=''\n",
    "    for i in who_drugs:\n",
    "        antibiogram += row[('PREDICTION',i)]\n",
    "    return antibiogram\n",
    "\n",
    "df1 = RAW_PREDICTIONS.unstack()\n",
    "df1['ANTIBIOGRAM'] = df1.apply(make_antibiogram, axis=1)\n",
    "\n",
    "PREDICTIONS.reset_index(inplace=True)\n",
    "df2 = PREDICTIONS[PREDICTIONS.SET=='nulls+minors']\n",
    "PREDICTIONS.set_index(['SET','ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "df2.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "df2=df2[['PREDICTION']].unstack()\n",
    "# df2=df2.droplevel(0, axis=1)\n",
    "df2['ANTIBIOGRAM'] = df2.apply(make_antibiogram, axis=1)\n",
    "\n",
    "mask = (df1.ANTIBIOGRAM != df2.ANTIBIOGRAM)\n",
    "assert len(df1[mask]) == 0, 'some samples have different antibiograms!'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having the (alphabetical) antibiogram is useful, so let's calculate it for all three scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SET1</th>\n",
       "      <th>ANTIBIOGRAM1</th>\n",
       "      <th>SET2</th>\n",
       "      <th>ANTIBIOGRAM2</th>\n",
       "      <th>SET3</th>\n",
       "      <th>ANTIBIOGRAM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR13286038</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286039</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSRSSSRSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286042</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SET1     ANTIBIOGRAM1   SET2     ANTIBIOGRAM2  \\\n",
       "ENA_RUN_ACCESSION                                                   \n",
       "ERR13286038        basic  RRRUSSRRSUSRRSS  nulls  RRRUSSRRSUSRRSS   \n",
       "ERR13286039        basic  RSSSSSSSSSSSSSS  nulls  RSSSSSSSSSSSSSS   \n",
       "ERR13286042        basic  RRRRSURRSSRRRRR  nulls  RRRRSURRSSRRRRR   \n",
       "\n",
       "                           SET3     ANTIBIOGRAM3  \n",
       "ENA_RUN_ACCESSION                                 \n",
       "ERR13286038        nulls+minors  RRRUSSRRSUSRRSS  \n",
       "ERR13286039        nulls+minors  RSSSRSSSRSSSSSS  \n",
       "ERR13286042        nulls+minors  RRRRSURRSSRRRRR  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = PREDICTIONS.unstack()\n",
    "df['ANTIBIOGRAM'] = df.apply(make_antibiogram, axis=1)\n",
    "df = df[[('ANTIBIOGRAM','')]]\n",
    "df = df.droplevel(1, axis=1)\n",
    "df.reset_index(inplace=True)\n",
    "df1 = df[df.SET=='basic']\n",
    "df2 = df[df.SET=='nulls']\n",
    "df3 = df[df.SET=='nulls+minors']\n",
    "df1.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df2.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df3.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df1 = df1.rename(columns={'SET':\"SET1\",'ANTIBIOGRAM':'ANTIBIOGRAM1'})\n",
    "df2 = df2.rename(columns={'SET':\"SET2\",'ANTIBIOGRAM':'ANTIBIOGRAM2'})\n",
    "df3 = df3.rename(columns={'SET':\"SET3\",'ANTIBIOGRAM':'ANTIBIOGRAM3'})\n",
    "ANTIBIOGRAMS = df1.join(df2).join(df3)\n",
    "ANTIBIOGRAMS.to_csv('dat/ANTIBIOGRAMS.csv')\n",
    "ANTIBIOGRAMS[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now look to see how many samples out of 2,663 change between scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calling nulls at resistance loci changes the antibiogram of 33 samples,whilst also calling minor alleles leads to 210 samples changing antibiogram\n"
     ]
    }
   ],
   "source": [
    "print(f\"Calling nulls at resistance loci changes the antibiogram of {(ANTIBIOGRAMS.ANTIBIOGRAM1!=ANTIBIOGRAMS.ANTIBIOGRAM2).sum()} samples,\\\n",
    "whilst also calling minor alleles leads to {(ANTIBIOGRAMS.ANTIBIOGRAM2!=ANTIBIOGRAMS.ANTIBIOGRAM3).sum()} samples changing antibiogram\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SET1</th>\n",
       "      <th>ANTIBIOGRAM1</th>\n",
       "      <th>SET2</th>\n",
       "      <th>ANTIBIOGRAM2</th>\n",
       "      <th>SET3</th>\n",
       "      <th>ANTIBIOGRAM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR13288807</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSSSSSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSSSSSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2510689</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSSSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSSSSSFRRFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSSSSSFRRFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2515400</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSUSSSRSSRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSUSSSRSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSUSSSRSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2515461</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSUUSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SFSSSSUUSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SFSSSSUUSSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2515574</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SFSSSSSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SFSSSSSSSSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4796971</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSUUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSSSFUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4797493</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SFSSSSSSSUSSUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SFSSSSSSSUSSUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4798823</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSSSSSSSUSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RFFFFSSSSSFUFFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RFRRFSSSSSFUFFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4812546</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4813514</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSUUSSSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSUUSSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSUUSSSFUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4813902</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSFSSSSSUSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSFSSSSSUSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4815604</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FSSSSSSSSSFSSFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FSSSSSSSSSFSSFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4816369</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSRRSSSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSRRSSSRRSF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSRRSSSRRSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4820776</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSRRSSSURSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSRRSSSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSRRSSSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828223</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSUSSSUSURSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSUSSSUSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSSUSSSUSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828270</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSUUSUU</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSUFSUU</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSSUFSUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828365</th>\n",
       "      <td>basic</td>\n",
       "      <td>USSSSUSSSSSRSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FFFSSUSSSSSRFSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FFFSSUSSSSSRFSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828428</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSUSSSSSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSUSSSSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSUSSSSSFUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828439</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828466</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSFSSFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSSSSSSSFRSFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828625</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SFSSSSSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SFSSSSSSSSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4829163</th>\n",
       "      <td>basic</td>\n",
       "      <td>URSSSSSSSSSUUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>URSSSSSSSSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>URSSSSSSSSSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4829526</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4830814</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSSSSSSSSRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSSSSSSSSSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSSSSSSSSSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4831746</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSRSSSSSSSRUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSRSSSSSSFRUFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSRSSSSSSFRUFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4831818</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSSSFSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSSSSFSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5917712</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSUSSRRS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSSSSUSFRRS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSSSSUSFRRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5917765</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSUSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSUSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSUSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR9119131</th>\n",
       "      <td>basic</td>\n",
       "      <td>SUSSSUSSSSSSUSU</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SUSSSUSSSSSFUSU</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SUSSSUSSSSSFUSU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR9121714</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSRSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSRSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSRSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1144784</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSSSURSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSFSSSSSURSF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSFSSSSSURSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1146278</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSSSFSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSFSSSSRRSRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1165572</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSRSSSSSSSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSRSSSSSSSRRSF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSRSSSSSSSRRSF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SET1     ANTIBIOGRAM1   SET2     ANTIBIOGRAM2  \\\n",
       "ENA_RUN_ACCESSION                                                   \n",
       "ERR13288807        basic  RRSSSSSSSSSSSSS  nulls  RRSSSSSSSSSFSSS   \n",
       "ERR2510689         basic  RRRRSSSSSSSRRSS  nulls  RRRRSSSSSSFRRFF   \n",
       "ERR2515400         basic  RRRRSUSSSRSSRSS  nulls  RRRRSUSSSRSFRSS   \n",
       "ERR2515461         basic  SSSSSSUUSSSSSSS  nulls  SFSSSSUUSSSSSSS   \n",
       "ERR2515574         basic  SSSSSSSSSSSSSSS  nulls  SFSSSSSSSSSSSSS   \n",
       "ERR4796971         basic  SSSSSSSSSSSUUSS  nulls  SSSSSSSSSSSFUSS   \n",
       "ERR4797493         basic  SSSSSSSSSUSSUSS  nulls  SFSSSSSSSUSSUSS   \n",
       "ERR4798823         basic  RSSSSSSSSSSUSSS  nulls  RFFFFSSSSSFUFFF   \n",
       "ERR4812546         basic  SSSSSSSSSUSSSSS  nulls  SSSSSSSSSUSFSSS   \n",
       "ERR4813514         basic  SSSSSSUUSSSSUSS  nulls  SSSSSSUUSSSFUSS   \n",
       "ERR4813902         basic  SSSSSSSSSUSSSSS  nulls  SSSFSSSSSUSSSSS   \n",
       "ERR4815604         basic  SSSSSSSSSSSSSSS  nulls  FSSSSSSSSSFSSFF   \n",
       "ERR4816369         basic  RRRRSSRRSSSRRSS  nulls  RRRRSSRRSSSRRSF   \n",
       "ERR4820776         basic  RRRRSSRRSSSURSS  nulls  RRRRSSRRSSSFRSS   \n",
       "ERR4828223         basic  RSSSSUSSSUSURSS  nulls  RSSSSUSSSUSFRSS   \n",
       "ERR4828270         basic  SSSSSSSSSSUUSUU  nulls  SSSSSSSSSSUFSUU   \n",
       "ERR4828365         basic  USSSSUSSSSSRSSS  nulls  FFFSSUSSSSSRFSS   \n",
       "ERR4828428         basic  SSSSSUSSSSSSUSS  nulls  SSSSSUSSSSSFUSS   \n",
       "ERR4828439         basic  SSSSSSSSSUSSSSS  nulls  SSSSSSSSSUSFSSS   \n",
       "ERR4828466         basic  SSSSSSSSSSSSSSS  nulls  SSSSSSSSSSFSSFF   \n",
       "ERR4828625         basic  SSSSSSSSSSSSSSS  nulls  SFSSSSSSSSSSSSS   \n",
       "ERR4829163         basic  URSSSSSSSSSUUSS  nulls  URSSSSSSSSSFUSS   \n",
       "ERR4829526         basic  SSSSSSSSSUSSSSS  nulls  SSSSSSSSSUSFSSS   \n",
       "ERR4830814         basic  RRSSSSSSSSSSRSS  nulls  RRSSSSSSSSSFRSS   \n",
       "ERR4831746         basic  RRSRSSSSSSSRUSS  nulls  RRSRSSSSSSFRUFF   \n",
       "ERR4831818         basic  RSSSSSSSSSSSSSS  nulls  RSSSSSSFSSSFSSS   \n",
       "ERR5917712         basic  RRRRSSSSSUSSRRS  nulls  RRRRSSSSSUSFRRS   \n",
       "ERR5917765         basic  SSSSSUSSSSSSSSS  nulls  SSSSSUSSSSSFSSS   \n",
       "ERR9119131         basic  SUSSSUSSSSSSUSU  nulls  SUSSSUSSSSSFUSU   \n",
       "ERR9121714         basic  SSSSSSSSSRSSSSS  nulls  SSSSSSSSSRSFSSS   \n",
       "SRR1144784         basic  RRRRSSSSSSSURSS  nulls  RRRRSFSSSSSURSF   \n",
       "SRR1146278         basic  RRSSSSSSSSSSSSS  nulls  RRSSSFSSSSSSSSS   \n",
       "SRR1165572         basic  RRSRSSSSSSSRRSS  nulls  RRSRSSSSSSSRRSF   \n",
       "\n",
       "                           SET3     ANTIBIOGRAM3  \n",
       "ENA_RUN_ACCESSION                                 \n",
       "ERR13288807        nulls+minors  RRSSSSSSSSSFSSS  \n",
       "ERR2510689         nulls+minors  RRRRSSSSSSFRRFF  \n",
       "ERR2515400         nulls+minors  RRRRSUSSSRSFRSS  \n",
       "ERR2515461         nulls+minors  SFSSSSUUSSSSSSS  \n",
       "ERR2515574         nulls+minors  SFSSSSSSSSSSSSS  \n",
       "ERR4796971         nulls+minors  SSSSSSSSSSSFUSS  \n",
       "ERR4797493         nulls+minors  SFSSSSSSSUSSUSS  \n",
       "ERR4798823         nulls+minors  RFRRFSSSSSFUFFF  \n",
       "ERR4812546         nulls+minors  SSSSSSSSSUSFSSS  \n",
       "ERR4813514         nulls+minors  SSSSSSUUSSSFUSS  \n",
       "ERR4813902         nulls+minors  SSSFSSSSSUSSSSS  \n",
       "ERR4815604         nulls+minors  FSSSSSSSSSFSSFF  \n",
       "ERR4816369         nulls+minors  RRRRSSRRSSSRRSF  \n",
       "ERR4820776         nulls+minors  RRRRSSRRSSSFRSS  \n",
       "ERR4828223         nulls+minors  RSSSSUSSSUSFRSS  \n",
       "ERR4828270         nulls+minors  SSSSSSSSSSUFSUU  \n",
       "ERR4828365         nulls+minors  FFFSSUSSSSSRFSS  \n",
       "ERR4828428         nulls+minors  SSSSSUSSSSSFUSS  \n",
       "ERR4828439         nulls+minors  SSSSSSSSSUSFSSS  \n",
       "ERR4828466         nulls+minors  RSSSSSSSSSFRSFF  \n",
       "ERR4828625         nulls+minors  SFSSSSSSSSSSSSS  \n",
       "ERR4829163         nulls+minors  URSSSSSSSSSFRSS  \n",
       "ERR4829526         nulls+minors  SSSSSSSSSUSFSSS  \n",
       "ERR4830814         nulls+minors  RRSSSSSSSSSFRSS  \n",
       "ERR4831746         nulls+minors  RRSRSSSSSSFRUFF  \n",
       "ERR4831818         nulls+minors  RSSSSSSFSSSFSSS  \n",
       "ERR5917712         nulls+minors  RRRRSSSSSUSFRRS  \n",
       "ERR5917765         nulls+minors  SSSSSUSSSSSFSSS  \n",
       "ERR9119131         nulls+minors  SUSSSUSSSSSFUSU  \n",
       "ERR9121714         nulls+minors  SSSSSSSSSRSFSSS  \n",
       "SRR1144784         nulls+minors  RRRRSFSSSSSURSF  \n",
       "SRR1146278         nulls+minors  RRRRSFSSSSRRSRR  \n",
       "SRR1165572         nulls+minors  RRSRSSSSSSSRRSF  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ANTIBIOGRAMS[ANTIBIOGRAMS.ANTIBIOGRAM1!=ANTIBIOGRAMS.ANTIBIOGRAM2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('UKMYC', 'INH'),\n",
       " ('UKMYC', 'RIF'),\n",
       " ('MGIT', 'PZA'),\n",
       " ('UKMYC', 'EMB'),\n",
       " ('MGIT', 'BDQ'),\n",
       " ('UKMYC', 'LZD'),\n",
       " ('UKMYC', 'MXF'),\n",
       " ('UKMYC', 'LEV'),\n",
       " ('UKMYC', 'CFZ'),\n",
       " ('UKMYC', 'DLM'),\n",
       " ('UKMYC', 'AMI'),\n",
       " ('MGIT', 'STM'),\n",
       " ('UKMYC', 'ETH'),\n",
       " ('UKMYC', 'KAN'),\n",
       " ('MGIT', 'CAP')]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_drug_order = pandas.read_csv('dat/drugs/results_drugs.csv')\n",
    "results_drug_order.set_index(['pDST method', 'drug'], inplace=True)\n",
    "list(results_drug_order.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3/3 [00:31<00:00, 10.45s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>sensitivity_sem</th>\n",
       "      <th>specificity</th>\n",
       "      <th>specificity_sem</th>\n",
       "      <th>PPV</th>\n",
       "      <th>PPV_sem</th>\n",
       "      <th>RR</th>\n",
       "      <th>SR</th>\n",
       "      <th>UR</th>\n",
       "      <th>RS</th>\n",
       "      <th>SS</th>\n",
       "      <th>US</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set</th>\n",
       "      <th>drug</th>\n",
       "      <th>method</th>\n",
       "      <th>dataset</th>\n",
       "      <th>quality</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">INH</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">UKMYC</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>91.240875</td>\n",
       "      <td>0.480167</td>\n",
       "      <td>95.395150</td>\n",
       "      <td>0.367757</td>\n",
       "      <td>95.206597</td>\n",
       "      <td>0.370676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>93.746796</td>\n",
       "      <td>0.336085</td>\n",
       "      <td>96.624866</td>\n",
       "      <td>0.352979</td>\n",
       "      <td>96.618858</td>\n",
       "      <td>0.342746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>91.200000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.600000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.397490</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>456.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>448.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>93.913043</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.412556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.428571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>432.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>906.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <th>UKMYC</th>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>93.729501</td>\n",
       "      <td>0.406094</td>\n",
       "      <td>95.957331</td>\n",
       "      <td>0.314197</td>\n",
       "      <td>95.445435</td>\n",
       "      <td>0.375315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">nulls+minors</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">KAN</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">UKMYC</th>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>HIGH</th>\n",
       "      <td>84.298262</td>\n",
       "      <td>0.713334</td>\n",
       "      <td>98.322163</td>\n",
       "      <td>0.199675</td>\n",
       "      <td>95.369505</td>\n",
       "      <td>0.574964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>76.470588</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.730606</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.086957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>221.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>650.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>84.615385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.543689</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>96.069869</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>220.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>880.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CAP</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">MGIT</th>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>74.983621</td>\n",
       "      <td>0.941786</td>\n",
       "      <td>98.295397</td>\n",
       "      <td>0.200364</td>\n",
       "      <td>94.980289</td>\n",
       "      <td>0.593964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>75.098814</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.254799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>831.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 sensitivity  sensitivity_sem  \\\n",
       "set          drug method dataset        quality                                 \n",
       "basic        INH  UKMYC  bootstrapped50 ALL        91.240875         0.480167   \n",
       "                                        HIGH       93.746796         0.336085   \n",
       "                         entire         ALL        91.200000         0.000000   \n",
       "                                        HIGH       93.913043         0.000000   \n",
       "             RIF  UKMYC  bootstrapped50 ALL        93.729501         0.406094   \n",
       "...                                                      ...              ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH       84.298262         0.713334   \n",
       "                         entire         ALL        76.470588         0.000000   \n",
       "                                        HIGH       84.615385         0.000000   \n",
       "             CAP  MGIT   bootstrapped50 ALL        74.983621         0.941786   \n",
       "                         entire         ALL        75.098814         0.000000   \n",
       "\n",
       "                                                 specificity  specificity_sem  \\\n",
       "set          drug method dataset        quality                                 \n",
       "basic        INH  UKMYC  bootstrapped50 ALL        95.395150         0.367757   \n",
       "                                        HIGH       96.624866         0.352979   \n",
       "                         entire         ALL        95.600000         0.000000   \n",
       "                                        HIGH       96.412556         0.000000   \n",
       "             RIF  UKMYC  bootstrapped50 ALL        95.957331         0.314197   \n",
       "...                                                      ...              ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH       98.322163         0.199675   \n",
       "                         entire         ALL        98.730606         0.000000   \n",
       "                                        HIGH       98.543689         0.000000   \n",
       "             CAP  MGIT   bootstrapped50 ALL        98.295397         0.200364   \n",
       "                         entire         ALL        98.254799         0.000000   \n",
       "\n",
       "                                                       PPV   PPV_sem     RR  \\\n",
       "set          drug method dataset        quality                               \n",
       "basic        INH  UKMYC  bootstrapped50 ALL      95.206597  0.370676    NaN   \n",
       "                                        HIGH     96.618858  0.342746    NaN   \n",
       "                         entire         ALL      95.397490  0.000000  456.0   \n",
       "                                        HIGH     96.428571  0.000000  432.0   \n",
       "             RIF  UKMYC  bootstrapped50 ALL      95.445435  0.375315    NaN   \n",
       "...                                                    ...       ...    ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH     95.369505  0.574964    NaN   \n",
       "                         entire         ALL      96.086957  0.000000  221.0   \n",
       "                                        HIGH     96.069869  0.000000  220.0   \n",
       "             CAP  MGIT   bootstrapped50 ALL      94.980289  0.593964    NaN   \n",
       "                         entire         ALL      95.000000  0.000000  190.0   \n",
       "\n",
       "                                                   SR    UR    RS     SS  \\\n",
       "set          drug method dataset        quality                            \n",
       "basic        INH  UKMYC  bootstrapped50 ALL       NaN   NaN   NaN    NaN   \n",
       "                                        HIGH      NaN   NaN   NaN    NaN   \n",
       "                         entire         ALL      33.0  11.0  22.0  448.0   \n",
       "                                        HIGH     23.0   5.0  16.0  404.0   \n",
       "             RIF  UKMYC  bootstrapped50 ALL       NaN   NaN   NaN    NaN   \n",
       "...                                               ...   ...   ...    ...   \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH      NaN   NaN   NaN    NaN   \n",
       "                         entire         ALL      56.0  12.0   9.0  650.0   \n",
       "                                        HIGH     34.0   6.0   9.0  563.0   \n",
       "             CAP  MGIT   bootstrapped50 ALL       NaN   NaN   NaN    NaN   \n",
       "                         entire         ALL      58.0   5.0  10.0  536.0   \n",
       "\n",
       "                                                   US   Total  \n",
       "set          drug method dataset        quality                \n",
       "basic        INH  UKMYC  bootstrapped50 ALL       NaN     NaN  \n",
       "                                        HIGH      NaN     NaN  \n",
       "                         entire         ALL      30.0  1000.0  \n",
       "                                        HIGH     26.0   906.0  \n",
       "             RIF  UKMYC  bootstrapped50 ALL       NaN     NaN  \n",
       "...                                               ...     ...  \n",
       "nulls+minors KAN  UKMYC  bootstrapped50 HIGH      NaN     NaN  \n",
       "                         entire         ALL      50.0  1000.0  \n",
       "                                        HIGH     46.0   880.0  \n",
       "             CAP  MGIT   bootstrapped50 ALL       NaN     NaN  \n",
       "                         entire         ALL      27.0   831.0  \n",
       "\n",
       "[156 rows x 13 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results={}\n",
    "PREDICTIONS.reset_index(inplace=True)\n",
    "PREDICTIONS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "\n",
    "for i in tqdm(['basic','nulls','nulls+minors']):\n",
    "    \n",
    "    predictions = PREDICTIONS[PREDICTIONS.SET==i]\n",
    "    df = copy.deepcopy(predictions.join(PHENOTYPES, how='inner'))\n",
    "\n",
    "    plate_df = df[df.PHENOTYPE_METHOD=='UKMYC']\n",
    "    plate_df.reset_index(inplace=True)\n",
    "    table = []\n",
    "    table = build_exact_table(table, plate_df, plate_drugs, ['HIGH', 'ALL'], i, 'UKMYC', include_fails=False)\n",
    "    table = build_bootstrapped_table(table,plate_df, plate_drugs, ['HIGH', 'ALL'], i, 'UKMYC', include_fails=False)\n",
    "\n",
    "    mgit_df = df[df.PHENOTYPE_METHOD=='MGIT']\n",
    "    mgit_df.reset_index(inplace=True)\n",
    "    table = build_exact_table(table, mgit_df, mgit_drugs, ['ALL'], i, 'MGIT', include_fails=False)\n",
    "    table = build_bootstrapped_table(table,mgit_df, mgit_drugs, ['ALL'], i, 'MGIT', include_fails=False)\n",
    "    results[i] = pandas.DataFrame(table, columns=['set','drug','method','quality','dataset','sensitivity','sensitivity_sem','specificity','specificity_sem','PPV','PPV_sem', 'RR', 'SR', 'UR', 'RS', 'SS', 'US', 'Total'])\n",
    "\n",
    "    results[i].set_index(['method','drug'],inplace=True)\n",
    "    results[i] = results[i][results[i].index.isin(list(results_drug_order.index))]\n",
    "    results[i].reset_index(inplace=True)\n",
    "    results[i].drug = results[i].drug.astype('category')\n",
    "    results[i].drug = results[i].drug.cat.set_categories(who_drugs)\n",
    "    results[i].sort_values(['drug', 'dataset', 'quality'], inplace=True)\n",
    "    results[i].set_index(['set','drug','method','dataset', 'quality'],inplace=True)\n",
    "\n",
    "results = pandas.concat(results.values())\n",
    "results.to_csv('dat/RESULTS.csv')\n",
    "results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the number of Fail results when we allow these rules to contribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 142 rows with a Fail call in the EFFECTS table (but these can be one variant affecting multiple drugs)\n",
      "Of the 142 Fails, 17 are duplicates\n",
      "And the most common genes are:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "rrs       82\n",
       "rpoB      19\n",
       "pncA       8\n",
       "embB       4\n",
       "gid        2\n",
       "Rv0678     2\n",
       "tlyA       2\n",
       "rplC       2\n",
       "ethA       1\n",
       "katG       1\n",
       "fabG1      1\n",
       "gyrB       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFFECTS.reset_index(inplace=True)\n",
    "EFFECTS['GENE_MUTATION'] = EFFECTS.GENE + '_' + EFFECTS.MUTATION\n",
    "print(f\"There are {len(EFFECTS[(EFFECTS.SET=='nulls') & (EFFECTS.PREDICTION=='F')])} rows with a Fail call in the EFFECTS table (but these can be one variant affecting multiple drugs)\")\n",
    "\n",
    "df = EFFECTS[(EFFECTS.SET=='nulls') & (EFFECTS.PREDICTION=='F')]\n",
    "df.set_index(['ENA_RUN_ACCESSION', 'GENE', 'MUTATION'], inplace=True)\n",
    "print(f\"Of the {len(df)} Fails, {df.index.duplicated().sum()} are duplicates\")\n",
    "\n",
    "df = df[~df.index.duplicated()]\n",
    "df.reset_index(inplace=True)\n",
    "print(\"And the most common genes are:\")\n",
    "df.GENE.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at which the RAVs most likely to be affected we can see how STM is affected more than AMI and KAN "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE_MUTATION\n",
       "rrs_c517x     36\n",
       "rrs_a514x     32\n",
       "rrs_c1402x     6\n",
       "rrs_a1401x     4\n",
       "rrs_g1484x     4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.GENE_MUTATION.value_counts()[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
