{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/tools/tools.py:6: UserWarning: A NumPy version >=1.23.5 and <2.3.0 is required for this version of SciPy (detected version 2.3.2)\n",
      "  import scipy.linalg\n"
     ]
    }
   ],
   "source": [
    "import pandas, copy\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "import statsmodels.api as sm\n",
    "\n",
    "from src.utils import build_exact_table, build_bootstrapped_table, create_predictions_table, plot_truthtables, plot_growth_boxplot\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "who_drugs = list(pandas.read_csv('dat/drugs/who2_drugs.csv').drug)\n",
    "plate_drugs = list(pandas.read_csv('dat/drugs/plate_drugs.csv').drug)\n",
    "other_drugs = list(pandas.read_csv('dat/drugs/other_drugs.csv').drug)\n",
    "mgit_drugs = list(pandas.read_csv('dat/drugs/mgit_drugs.csv').drug)\n",
    "\n",
    "drug_names_table = pandas.read_csv(\"dat/drugs/drug_names_lookup.csv\")\n",
    "drug_names_table.set_index(\"DRUG\", inplace=True)\n",
    "drug_names_lookup = {}\n",
    "for idx, row in drug_names_table.iterrows():\n",
    "    drug_names_lookup[idx] = row.DRUG_NAME.capitalize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>PHENOTYPE_METHOD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>site.10.subj.LA00559418.lab.LA00559418.iso.1</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGIT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            UNIQUEID  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                 \n",
       "ERR13286038       BDQ   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "                  LZD   site.10.subj.LA00559418.lab.LA00559418.iso.1   \n",
       "\n",
       "                       BINARY_PHENOTYPE PHENOTYPE_QUALITY PHENOTYPE_METHOD  \n",
       "ENA_RUN_ACCESSION DRUG                                                      \n",
       "ERR13286038       BDQ                 R               NaN             MGIT  \n",
       "                  LZD                 S               NaN             MGIT  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PHENOTYPES = pandas.read_csv('dat/PHENOTYPES.csv')\n",
    "PHENOTYPES.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "PHENOTYPES.sort_index(inplace=True)\n",
    "PHENOTYPES[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR4799790</th>\n",
       "      <th>AMI</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       PREDICTION\n",
       "ENA_RUN_ACCESSION DRUG           \n",
       "ERR4799790        AMI           R\n",
       "                  BDQ           S\n",
       "                  CAP           R"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_PREDICTIONS = pandas.read_csv('dat/RAW_PREDICTIONS.csv')\n",
    "RAW_PREDICTIONS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "RAW_PREDICTIONS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 503 null calls and 3118 minor allele calls in the main effects table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">ERR13286038</th>\n",
       "      <th>BDQ</th>\n",
       "      <td>Rv0678</td>\n",
       "      <td>138_ins_g:38</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>Rv0678</td>\n",
       "      <td>141_ins_c:60</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          GENE      MUTATION PREDICTION  EPISTASIS  \\\n",
       "ENA_RUN_ACCESSION DRUG                                               \n",
       "ERR13286038       BDQ   Rv0678  138_ins_g:38          R      False   \n",
       "                  BDQ   Rv0678  141_ins_c:60          R      False   \n",
       "                  CAP     tlyA          L11L          S      False   \n",
       "\n",
       "                        IS_MINOR_ALLELE  IS_NULL  \n",
       "ENA_RUN_ACCESSION DRUG                            \n",
       "ERR13286038       BDQ              True    False  \n",
       "                  BDQ              True    False  \n",
       "                  CAP             False    False  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS = pandas.read_csv('dat/RAW_EFFECTS.csv')\n",
    "RAW_EFFECTS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "RAW_EFFECTS.sort_index(inplace=True)\n",
    "print(f\"There are {RAW_EFFECTS.IS_NULL.sum()} null calls and {RAW_EFFECTS.IS_MINOR_ALLELE.sum()} minor allele calls in the main effects table\")\n",
    "RAW_EFFECTS[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most of the null calls are in the *rrs* gene so will affect the aminoglycosides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "rrs       424\n",
       "rpoB       20\n",
       "katG       17\n",
       "pncA       10\n",
       "embB        9\n",
       "Rv0678      8\n",
       "eis         4\n",
       "gid         2\n",
       "fabG1       2\n",
       "gyrB        2\n",
       "tlyA        2\n",
       "rplC        2\n",
       "ethA        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS[RAW_EFFECTS.IS_NULL].GENE.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whilst there are over 50 minor allele calls in genes that could affect resistance prediction for the fluorouinolones, rifampicin, bedaquline, pyrazinamide and ethambutol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "Rv0678    666\n",
       "embB      501\n",
       "rrs       368\n",
       "gyrA      368\n",
       "rpoB      169\n",
       "gyrB      132\n",
       "Rv2983    121\n",
       "ethA      106\n",
       "gid       104\n",
       "fabG1      90\n",
       "pncA       89\n",
       "katG       88\n",
       "rrl        53\n",
       "eis        52\n",
       "fgd1       43\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAW_EFFECTS[RAW_EFFECTS.IS_MINOR_ALLELE].GENE.value_counts()[:15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now build three EFFECTS tables that progressively include these nuances and from these build three PREDICTIONS tables that we can analyse by joining to PHENOTYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 51842/51842 [00:01<00:00, 33852.55it/s]\n",
      "100%|██████████| 52345/52345 [00:01<00:00, 31989.37it/s]\n",
      "100%|██████████| 55463/55463 [00:01<00:00, 34359.78it/s]\n"
     ]
    }
   ],
   "source": [
    "EFFECTS={}\n",
    "EFFECTS[1] = copy.deepcopy(RAW_EFFECTS[~(RAW_EFFECTS.IS_MINOR_ALLELE) & ~(RAW_EFFECTS.IS_NULL)])\n",
    "EFFECTS[1]['SET'] = 'basic'\n",
    "EFFECTS[2] = copy.deepcopy(RAW_EFFECTS[~(RAW_EFFECTS.IS_MINOR_ALLELE)])\n",
    "EFFECTS[2]['SET'] = 'nulls'\n",
    "EFFECTS[3] = copy.deepcopy(RAW_EFFECTS)\n",
    "EFFECTS[3]['SET'] = 'nulls+minors'\n",
    "assert len(EFFECTS[3]) >  len(EFFECTS[2]) >  len(EFFECTS[1])\n",
    "\n",
    "PREDICTIONS={}\n",
    "for i in [1,2,3]:\n",
    "    PREDICTIONS[i] = pandas.DataFrame(create_predictions_table(EFFECTS[i], who_drugs))\n",
    "\n",
    "assert len(RAW_PREDICTIONS) == len(PREDICTIONS[1]) == len(PREDICTIONS[2]) == len(PREDICTIONS[3])\n",
    "\n",
    "EFFECTS = pandas.concat(EFFECTS.values())\n",
    "EFFECTS.reset_index(inplace=True)\n",
    "EFFECTS.set_index(['SET','ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "EFFECTS.to_csv('dat/EFFECTS.csv')\n",
    "\n",
    "PREDICTIONS= pandas.concat(PREDICTIONS.values())\n",
    "PREDICTIONS.to_csv('dat/PREDICTIONS.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These tables each contain three scenarios (i) `basic` which ignores all null and minor alleles and is therefore closest to the WHOv2 catalogue as written, (ii) `nulls` which makes an `F` call at genetic loci associated with resistance where there are insufficient reads (two or fewer) and (iii) `nulls+minors` which adds in calls where there are three or more reads supporting a resistance-associated variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>CAP</th>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>fgd1</td>\n",
       "      <td>K270M</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              GENE MUTATION PREDICTION  EPISTASIS  \\\n",
       "SET   ENA_RUN_ACCESSION DRUG                                        \n",
       "basic ERR13286038       CAP   tlyA     L11L          S      False   \n",
       "                        DLM   fgd1    K270M          U      False   \n",
       "\n",
       "                              IS_MINOR_ALLELE  IS_NULL  \n",
       "SET   ENA_RUN_ACCESSION DRUG                            \n",
       "basic ERR13286038       CAP             False    False  \n",
       "                        DLM             False    False  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFFECTS[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             PREDICTION\n",
       "SET   ENA_RUN_ACCESSION DRUG           \n",
       "basic ERR13286038       INH           R\n",
       "                        RIF           R"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PREDICTIONS[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check that has all worked, first by picking a sample and looking at its rows in an EFFECTS table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>SET</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EPISTASIS</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>IS_NULL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"22\" valign=\"top\">ERR13286038</th>\n",
       "      <th>CAP</th>\n",
       "      <td>nulls</td>\n",
       "      <td>tlyA</td>\n",
       "      <td>L11L</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fgd1</td>\n",
       "      <td>K270M</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>nulls</td>\n",
       "      <td>embB</td>\n",
       "      <td>A409P</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>nulls</td>\n",
       "      <td>embB</td>\n",
       "      <td>E405D</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>ethA</td>\n",
       "      <td>Y211S</td>\n",
       "      <td>U</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>c-15t</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>fabG1</td>\n",
       "      <td>c-15t</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INH</th>\n",
       "      <td>nulls</td>\n",
       "      <td>katG</td>\n",
       "      <td>S315T</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>D94G</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>E21Q</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>G668D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>S95T</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>nulls</td>\n",
       "      <td>pncA</td>\n",
       "      <td>H71Y</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>D103D</td>\n",
       "      <td>S</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>nulls</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>H445Y</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STM</th>\n",
       "      <td>nulls</td>\n",
       "      <td>gid</td>\n",
       "      <td>A134E</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SET   GENE MUTATION PREDICTION  EPISTASIS  \\\n",
       "ENA_RUN_ACCESSION DRUG                                                \n",
       "ERR13286038       CAP   nulls   tlyA     L11L          S      False   \n",
       "                  DLM   nulls   fgd1    K270M          U      False   \n",
       "                  EMB   nulls   embB    A409P          U      False   \n",
       "                  EMB   nulls   embB    E405D          U      False   \n",
       "                  ETH   nulls   ethA    Y211S          U      False   \n",
       "                  ETH   nulls  fabG1    c-15t          R      False   \n",
       "                  INH   nulls  fabG1    c-15t          R      False   \n",
       "                  INH   nulls   katG    S315T          R      False   \n",
       "                  LEV   nulls   gyrA     A90V          R      False   \n",
       "                  LEV   nulls   gyrA     D94G          R      False   \n",
       "                  LEV   nulls   gyrA     E21Q          S      False   \n",
       "                  LEV   nulls   gyrA    G668D          S      False   \n",
       "                  LEV   nulls   gyrA     S95T          S      False   \n",
       "                  MXF   nulls   gyrA     A90V          R      False   \n",
       "                  MXF   nulls   gyrA     D94G          R      False   \n",
       "                  MXF   nulls   gyrA     E21Q          S      False   \n",
       "                  MXF   nulls   gyrA    G668D          S      False   \n",
       "                  MXF   nulls   gyrA     S95T          S      False   \n",
       "                  PZA   nulls   pncA     H71Y          R      False   \n",
       "                  RIF   nulls   rpoB    D103D          S      False   \n",
       "                  RIF   nulls   rpoB    H445Y          R      False   \n",
       "                  STM   nulls    gid    A134E          R      False   \n",
       "\n",
       "                        IS_MINOR_ALLELE  IS_NULL  \n",
       "ENA_RUN_ACCESSION DRUG                            \n",
       "ERR13286038       CAP             False    False  \n",
       "                  DLM             False    False  \n",
       "                  EMB             False    False  \n",
       "                  EMB             False    False  \n",
       "                  ETH             False    False  \n",
       "                  ETH             False    False  \n",
       "                  INH             False    False  \n",
       "                  INH             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  LEV             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  MXF             False    False  \n",
       "                  PZA             False    False  \n",
       "                  RIF             False    False  \n",
       "                  RIF             False    False  \n",
       "                  STM             False    False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_sample = 'ERR13286038'\n",
    "effects = copy.deepcopy(EFFECTS)\n",
    "effects.reset_index(inplace=True)\n",
    "effects = effects[(effects.SET=='nulls') & (effects.ENA_RUN_ACCESSION==check_sample)]\n",
    "effects.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now compare that to the calculated per drug predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:00<00:00, 31600.92it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PREDICTION</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SET</th>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th>DRUG</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"15\" valign=\"top\">nulls</th>\n",
       "      <th rowspan=\"15\" valign=\"top\">ERR13286038</th>\n",
       "      <th>INH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PZA</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMB</th>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BDQ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LZD</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXF</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEV</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFZ</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DLM</th>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMI</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STM</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETH</th>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KAN</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAP</th>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             PREDICTION\n",
       "SET   ENA_RUN_ACCESSION DRUG           \n",
       "nulls ERR13286038       INH           R\n",
       "                        RIF           R\n",
       "                        PZA           R\n",
       "                        EMB           U\n",
       "                        BDQ           S\n",
       "                        LZD           S\n",
       "                        MXF           R\n",
       "                        LEV           R\n",
       "                        CFZ           S\n",
       "                        DLM           U\n",
       "                        AMI           S\n",
       "                        STM           R\n",
       "                        ETH           R\n",
       "                        KAN           S\n",
       "                        CAP           S"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_predictions_table(effects, who_drugs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we have a `RAW_PREDICTIONS` table parsed directly from the individual sample `json` files we can compare `PREDICTIONS['nulls+minors']` to that and they should be identical.\n",
    "\n",
    "A direct way of doing that is to make an antibiogram for each sample (which will automatically be in alphabetical order of drug), and then look for samples where they are different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_antibiogram(row):\n",
    "    antibiogram=''\n",
    "    for i in who_drugs:\n",
    "        antibiogram += row[('PREDICTION',i)]\n",
    "    return antibiogram\n",
    "\n",
    "df1 = RAW_PREDICTIONS.unstack()\n",
    "df1['ANTIBIOGRAM'] = df1.apply(make_antibiogram, axis=1)\n",
    "\n",
    "PREDICTIONS.reset_index(inplace=True)\n",
    "df2 = PREDICTIONS[PREDICTIONS.SET=='nulls+minors']\n",
    "PREDICTIONS.set_index(['SET','ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "df2.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "df2=df2[['PREDICTION']].unstack()\n",
    "# df2=df2.droplevel(0, axis=1)\n",
    "df2['ANTIBIOGRAM'] = df2.apply(make_antibiogram, axis=1)\n",
    "\n",
    "mask = (df1.ANTIBIOGRAM != df2.ANTIBIOGRAM)\n",
    "assert len(df1[mask]) == 0, 'some samples have different antibiograms!'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having the (alphabetical) antibiogram is useful, so let's calculate it for all three scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SET1</th>\n",
       "      <th>ANTIBIOGRAM1</th>\n",
       "      <th>SET2</th>\n",
       "      <th>ANTIBIOGRAM2</th>\n",
       "      <th>SET3</th>\n",
       "      <th>ANTIBIOGRAM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR13286038</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRUSSRRSUSRRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRURSRRRUSRRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286039</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSRSSSRSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR13286042</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSURRSSRRRRR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SET1     ANTIBIOGRAM1   SET2     ANTIBIOGRAM2  \\\n",
       "ENA_RUN_ACCESSION                                                   \n",
       "ERR13286038        basic  RRRUSSRRSUSRRSS  nulls  RRRUSSRRSUSRRSS   \n",
       "ERR13286039        basic  RSSSSSSSSSSSSSS  nulls  RSSSSSSSSSSSSSS   \n",
       "ERR13286042        basic  RRRRSURRSSRRRRR  nulls  RRRRSURRSSRRRRR   \n",
       "\n",
       "                           SET3     ANTIBIOGRAM3  \n",
       "ENA_RUN_ACCESSION                                 \n",
       "ERR13286038        nulls+minors  RRRURSRRRUSRRSS  \n",
       "ERR13286039        nulls+minors  RSSSRSSSRSSSSSS  \n",
       "ERR13286042        nulls+minors  RRRRSURRSSRRRRR  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = PREDICTIONS.unstack()\n",
    "df['ANTIBIOGRAM'] = df.apply(make_antibiogram, axis=1)\n",
    "df = df[[('ANTIBIOGRAM','')]]\n",
    "df = df.droplevel(1, axis=1)\n",
    "df.reset_index(inplace=True)\n",
    "df1 = df[df.SET=='basic']\n",
    "df2 = df[df.SET=='nulls']\n",
    "df3 = df[df.SET=='nulls+minors']\n",
    "df1.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df2.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df3.set_index('ENA_RUN_ACCESSION', inplace=True)\n",
    "df1 = df1.rename(columns={'SET':\"SET1\",'ANTIBIOGRAM':'ANTIBIOGRAM1'})\n",
    "df2 = df2.rename(columns={'SET':\"SET2\",'ANTIBIOGRAM':'ANTIBIOGRAM2'})\n",
    "df3 = df3.rename(columns={'SET':\"SET3\",'ANTIBIOGRAM':'ANTIBIOGRAM3'})\n",
    "ANTIBIOGRAMS = df1.join(df2).join(df3)\n",
    "ANTIBIOGRAMS.to_csv('dat/ANTIBIOGRAMS.csv')\n",
    "ANTIBIOGRAMS[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now look to see how many samples out of 2,663 change between scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calling nulls at resistance loci changes the antibiogram of 44 samples,whilst also calling minor alleles leads to 375 samples changing antibiogram\n"
     ]
    }
   ],
   "source": [
    "print(f\"Calling nulls at resistance loci changes the antibiogram of {(ANTIBIOGRAMS.ANTIBIOGRAM1!=ANTIBIOGRAMS.ANTIBIOGRAM2).sum()} samples,\\\n",
    "whilst also calling minor alleles leads to {(ANTIBIOGRAMS.ANTIBIOGRAM2!=ANTIBIOGRAMS.ANTIBIOGRAM3).sum()} samples changing antibiogram\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SET1</th>\n",
       "      <th>ANTIBIOGRAM1</th>\n",
       "      <th>SET2</th>\n",
       "      <th>ANTIBIOGRAM2</th>\n",
       "      <th>SET3</th>\n",
       "      <th>ANTIBIOGRAM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENA_RUN_ACCESSION</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR13288807</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSSSSSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSSSSSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2510689</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSSSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSSSSSFRRFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSSSSSFRRFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2515400</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSUSSSSSSRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSUSSSSSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSUSSSRSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4797493</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SFSSSSSSSUFSUFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SFSSSSSSSUFSUFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4797495</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4797833</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4798658</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSRRSSRURRR</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSRRSSRFRRR</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSRRSSRFRRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4798823</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSUSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FFFFFSSSSSFFFFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FFRRFSSSSSFFFFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4811314</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSUUSSSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FSSSSSUUSSSSUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FRSSSSUUSSSSUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4813902</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSFSSSSSUSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSFSSSSSUSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4814090</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSUSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSFSUSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSFSUSSSSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4815604</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FSSSSSSSSSFSSFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FSSSSSSSSSFSSFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4816369</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSRRSSSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSRRSSFRRFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSRRSSFRRFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4817820</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSUSSRSU</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSSSSUSFRSU</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSSSSUSFRSU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4821059</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSUSSRSU</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSSSSUSFRSU</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSSSSUSFRSU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4822360</th>\n",
       "      <td>basic</td>\n",
       "      <td>SRSRSSSSSSRRSRR</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FRSRSSSSSSRRSRR</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FRSRSSSSSSRRSRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828223</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSUSSSUSURSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSUSSSUSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSSUSSSUSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828365</th>\n",
       "      <td>basic</td>\n",
       "      <td>USSSSUSSSSSRSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FFFSSUSSSSSRFSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FFFSSUSSSSSRFSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828374</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828403</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828428</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSUSSSSSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSUSSSSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSUSSSSSFUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828439</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4828466</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSSFFSFF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSSSSSSSFRSFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4829000</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSRUSSSUSSUSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSRUSSSUSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSRUSSSUSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4829167</th>\n",
       "      <td>basic</td>\n",
       "      <td>SRSRRSRRRSUUSUS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FRSRRSRRRSUUSUS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FRSRRSRRRSUUSUS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4829526</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4830745</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRSSSSSSSSSRRS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRSSSSSSSSFRRS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRSSSSSSSSFRRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4830814</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSSSSSSSSRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSSSSSSSSSFRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSSSSSSSSSFRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4831746</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSRSSSSSSSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSRSSSSSSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSRSSSSSSSFUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4831818</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSSSSFSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSSSSFSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4831890</th>\n",
       "      <td>basic</td>\n",
       "      <td>RSSSUSSSUSSSUSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RSSSUSSSUSSFUSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RSSSUSSSUSSFUSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5917712</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSUSSRRS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRSSSSSUSFRRS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRSSSSSUSFRRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5917762</th>\n",
       "      <td>basic</td>\n",
       "      <td>USSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>USSSSSSSSUSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>USSSSSSSSUSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5917765</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSUSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSUSSSSSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSUSSSSSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR8975151</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSFSSSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSFSSSSSSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR8975158</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSUSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSFSSSSSUSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSFSSSSSUSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR9119131</th>\n",
       "      <td>basic</td>\n",
       "      <td>SUSSSUSSSSSSUSU</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SUSSSUSSSSSFUSU</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SUSSSUSSSSSFUSU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR9121714</th>\n",
       "      <td>basic</td>\n",
       "      <td>SSSSSSSSSRSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>SSSSSSSSSRSFSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>SSSSSSSSSRSFSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR9992649</th>\n",
       "      <td>basic</td>\n",
       "      <td>SRSSRSSSRSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FRSSRSSSRSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FRSSRSSSRSSSSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR9992939</th>\n",
       "      <td>basic</td>\n",
       "      <td>SRSRRSRRRSSSRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>FRSRRSRRRSSSRSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>FRSRRSRRRSSSRSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1144784</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRRRSSSSSSSURSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRRRFFSSSSSURSF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRFFSSSSSURSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1144788</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSUSSSUUUSUU</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSFSUSSSUUUSUU</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSFSUSSSUUUSUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1146278</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSSSSSSSSSSSSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSFSFSSSSSSSSS</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRRRRFSSRSRRSRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1165572</th>\n",
       "      <td>basic</td>\n",
       "      <td>RRSRSSSSSSSRRSS</td>\n",
       "      <td>nulls</td>\n",
       "      <td>RRSRSSSSSSSRRSF</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RRSRSSSSSSSRRSF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SET1     ANTIBIOGRAM1   SET2     ANTIBIOGRAM2  \\\n",
       "ENA_RUN_ACCESSION                                                   \n",
       "ERR13288807        basic  RRSSSSSSSSSSSSS  nulls  RRSSSSSSSSSFSSS   \n",
       "ERR2510689         basic  RRRRSSSSSSSRRSS  nulls  RRRRSSSSSSFRRFF   \n",
       "ERR2515400         basic  RRRRSUSSSSSSRSS  nulls  RRRRSUSSSSSFRSS   \n",
       "ERR4797493         basic  SSSSSSSSSUSSUSS  nulls  SFSSSSSSSUFSUFF   \n",
       "ERR4797495         basic  SSSSSSSSSUSSSSS  nulls  SSSSSSSSSUSFSSS   \n",
       "ERR4797833         basic  SSSSSSSSSSSSSSS  nulls  SSSSSSSSSSSFSSS   \n",
       "ERR4798658         basic  RRRRSSRRSSRURRR  nulls  RRRRSSRRSSRFRRR   \n",
       "ERR4798823         basic  SSSSSSSSSSSUSSS  nulls  FFFFFSSSSSFFFFF   \n",
       "ERR4811314         basic  SSSSSSUUSSSSUSS  nulls  FSSSSSUUSSSSUSS   \n",
       "ERR4813902         basic  SSSSSSSSSUSSSSS  nulls  SSSFSSSSSUSSSSS   \n",
       "ERR4814090         basic  SSSSSUSSSSSSSSS  nulls  SSSFSUSSSSSSSSS   \n",
       "ERR4815604         basic  SSSSSSSSSSSSSSS  nulls  FSSSSSSSSSFSSFF   \n",
       "ERR4816369         basic  RRRRSSRRSSSRRSS  nulls  RRRRSSRRSSFRRFF   \n",
       "ERR4817820         basic  RRRRSSSSSUSSRSU  nulls  RRRRSSSSSUSFRSU   \n",
       "ERR4821059         basic  RRRRSSSSSUSSRSU  nulls  RRRRSSSSSUSFRSU   \n",
       "ERR4822360         basic  SRSRSSSSSSRRSRR  nulls  FRSRSSSSSSRRSRR   \n",
       "ERR4828223         basic  RSSSSUSSSUSURSS  nulls  RSSSSUSSSUSFRSS   \n",
       "ERR4828365         basic  USSSSUSSSSSRSSS  nulls  FFFSSUSSSSSRFSS   \n",
       "ERR4828374         basic  SSSSSSSSSSSSSSS  nulls  SSSSSSSSSSSFSSS   \n",
       "ERR4828403         basic  SSSSSSSSSSSSSSS  nulls  SSSSSSSSSSSFSSS   \n",
       "ERR4828428         basic  SSSSSUSSSSSSUSS  nulls  SSSSSUSSSSSFUSS   \n",
       "ERR4828439         basic  SSSSSSSSSUSSSSS  nulls  SSSSSSSSSUSFSSS   \n",
       "ERR4828466         basic  SSSSSSSSSSSSSSS  nulls  SSSSSSSSSSFFSFF   \n",
       "ERR4829000         basic  RRSRUSSSUSSUSSS  nulls  RRSRUSSSUSSFSSS   \n",
       "ERR4829167         basic  SRSRRSRRRSUUSUS  nulls  FRSRRSRRRSUUSUS   \n",
       "ERR4829526         basic  SSSSSSSSSUSSSSS  nulls  SSSSSSSSSUSFSSS   \n",
       "ERR4830745         basic  RRRSSSSSSSSSRRS  nulls  RRRSSSSSSSSFRRS   \n",
       "ERR4830814         basic  RRSSSSSSSSSSRSS  nulls  RRSSSSSSSSSFRSS   \n",
       "ERR4831746         basic  RRSRSSSSSSSSUSS  nulls  RRSRSSSSSSSFUSS   \n",
       "ERR4831818         basic  RSSSSSSSSSSSSSS  nulls  RSSSSSSFSSSFSSS   \n",
       "ERR4831890         basic  RSSSUSSSUSSSUSS  nulls  RSSSUSSSUSSFUSS   \n",
       "ERR5917712         basic  RRRRSSSSSUSSRRS  nulls  RRRRSSSSSUSFRRS   \n",
       "ERR5917762         basic  USSSSSSSSUSSSSS  nulls  USSSSSSSSUSFSSS   \n",
       "ERR5917765         basic  SSSSSUSSSSSSSSS  nulls  SSSSSUSSSSSFSSS   \n",
       "ERR8975151         basic  SSSSSSSSSSSSSSS  nulls  SSSFSSSSSSSSSSS   \n",
       "ERR8975158         basic  SSSSSSSSSUSSSSS  nulls  SSSFSSSSSUSSSSS   \n",
       "ERR9119131         basic  SUSSSUSSSSSSUSU  nulls  SUSSSUSSSSSFUSU   \n",
       "ERR9121714         basic  SSSSSSSSSRSSSSS  nulls  SSSSSSSSSRSFSSS   \n",
       "ERR9992649         basic  SRSSRSSSRSSSSSS  nulls  FRSSRSSSRSSSSSS   \n",
       "ERR9992939         basic  SRSRRSRRRSSSRSS  nulls  FRSRRSRRRSSSRSS   \n",
       "SRR1144784         basic  RRRRSSSSSSSURSS  nulls  RRRRFFSSSSSURSF   \n",
       "SRR1144788         basic  RRSSSUSSSUUUSUU  nulls  RRSFSUSSSUUUSUU   \n",
       "SRR1146278         basic  RRSSSSSSSSSSSSS  nulls  RRSFSFSSSSSSSSS   \n",
       "SRR1165572         basic  RRSRSSSSSSSRRSS  nulls  RRSRSSSSSSSRRSF   \n",
       "\n",
       "                           SET3     ANTIBIOGRAM3  \n",
       "ENA_RUN_ACCESSION                                 \n",
       "ERR13288807        nulls+minors  RRSSSSSSSSSFSSS  \n",
       "ERR2510689         nulls+minors  RRRRSSSSSSFRRFF  \n",
       "ERR2515400         nulls+minors  RRRRSUSSSRSFRSS  \n",
       "ERR4797493         nulls+minors  SFSSSSSSSUFSUFF  \n",
       "ERR4797495         nulls+minors  SSSSSSSSSUSFSSS  \n",
       "ERR4797833         nulls+minors  SSSSSSSSSSSFSSS  \n",
       "ERR4798658         nulls+minors  RRRRSSRRSSRFRRR  \n",
       "ERR4798823         nulls+minors  FFRRFSSSSSFFFFF  \n",
       "ERR4811314         nulls+minors  FRSSSSUUSSSSUSS  \n",
       "ERR4813902         nulls+minors  SSSFSSSSSUSSSSS  \n",
       "ERR4814090         nulls+minors  SSSFSUSSSSSSSSS  \n",
       "ERR4815604         nulls+minors  FSSSSSSSSSFSSFF  \n",
       "ERR4816369         nulls+minors  RRRRSSRRSSFRRFF  \n",
       "ERR4817820         nulls+minors  RRRRSSSSSUSFRSU  \n",
       "ERR4821059         nulls+minors  RRRRSSSSSUSFRSU  \n",
       "ERR4822360         nulls+minors  FRSRSSSSSSRRSRR  \n",
       "ERR4828223         nulls+minors  RSSSSUSSSUSFRSS  \n",
       "ERR4828365         nulls+minors  FFFSSUSSSSSRFSS  \n",
       "ERR4828374         nulls+minors  SSSSSSSSSSSFSSS  \n",
       "ERR4828403         nulls+minors  SSSSSSSSSSSFSSS  \n",
       "ERR4828428         nulls+minors  SSSSSUSSSSSFUSS  \n",
       "ERR4828439         nulls+minors  SSSSSSSSSUSFSSS  \n",
       "ERR4828466         nulls+minors  RSSSSSSSSSFRSFF  \n",
       "ERR4829000         nulls+minors  RRSRUSSSUSSFSSS  \n",
       "ERR4829167         nulls+minors  FRSRRSRRRSUUSUS  \n",
       "ERR4829526         nulls+minors  SSSSSSSSSUSFSSS  \n",
       "ERR4830745         nulls+minors  RRRSSSSSSSSFRRS  \n",
       "ERR4830814         nulls+minors  RRSSSSSSSSSFRSS  \n",
       "ERR4831746         nulls+minors  RRSRSSSSSSSFUSS  \n",
       "ERR4831818         nulls+minors  RSSSSSSFSSSFSSS  \n",
       "ERR4831890         nulls+minors  RSSSUSSSUSSFUSS  \n",
       "ERR5917712         nulls+minors  RRRRSSSSSUSFRRS  \n",
       "ERR5917762         nulls+minors  USSSSSSSSUSFSSS  \n",
       "ERR5917765         nulls+minors  SSSSSUSSSSSFSSS  \n",
       "ERR8975151         nulls+minors  SSSFSSSSSSSSSSS  \n",
       "ERR8975158         nulls+minors  SSSFSSSSSUSSSSS  \n",
       "ERR9119131         nulls+minors  SUSSSUSSSSSFUSU  \n",
       "ERR9121714         nulls+minors  SSSSSSSSSRSFSSS  \n",
       "ERR9992649         nulls+minors  FRSSRSSSRSSSSSS  \n",
       "ERR9992939         nulls+minors  FRSRRSRRRSSSRSS  \n",
       "SRR1144784         nulls+minors  RRRRFFSSSSSURSF  \n",
       "SRR1144788         nulls+minors  RRSFSUSSSUUUSUU  \n",
       "SRR1146278         nulls+minors  RRRRRFSSRSRRSRR  \n",
       "SRR1165572         nulls+minors  RRSRSSSSSSSRRSF  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ANTIBIOGRAMS[ANTIBIOGRAMS.ANTIBIOGRAM1!=ANTIBIOGRAMS.ANTIBIOGRAM2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('UKMYC', 'INH'),\n",
       " ('UKMYC', 'RIF'),\n",
       " ('MGIT', 'PZA'),\n",
       " ('UKMYC', 'EMB'),\n",
       " ('MGIT', 'BDQ'),\n",
       " ('UKMYC', 'LZD'),\n",
       " ('UKMYC', 'MXF'),\n",
       " ('UKMYC', 'LEV'),\n",
       " ('UKMYC', 'CFZ'),\n",
       " ('UKMYC', 'DLM'),\n",
       " ('UKMYC', 'AMI'),\n",
       " ('MGIT', 'STM'),\n",
       " ('UKMYC', 'ETH'),\n",
       " ('UKMYC', 'KAN'),\n",
       " ('MGIT', 'CAP')]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_drug_order = pandas.read_csv('dat/drugs/results_drugs.csv')\n",
    "results_drug_order.set_index(['pDST method', 'drug'], inplace=True)\n",
    "list(results_drug_order.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4/4 [00:33<00:00,  8.41s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>sensitivity_sem</th>\n",
       "      <th>specificity</th>\n",
       "      <th>specificity_sem</th>\n",
       "      <th>PPV</th>\n",
       "      <th>PPV_sem</th>\n",
       "      <th>RR</th>\n",
       "      <th>SR</th>\n",
       "      <th>UR</th>\n",
       "      <th>RS</th>\n",
       "      <th>SS</th>\n",
       "      <th>US</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set</th>\n",
       "      <th>drug</th>\n",
       "      <th>method</th>\n",
       "      <th>dataset</th>\n",
       "      <th>quality</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">basic</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">INH</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">UKMYC</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">bootstrap-0</th>\n",
       "      <th>ALL</th>\n",
       "      <td>88.803089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.020747</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.041322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>93.172691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.414343</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.265560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">bootstrap-1</th>\n",
       "      <th>ALL</th>\n",
       "      <td>91.828794</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94.650206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94.779116</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIGH</th>\n",
       "      <td>94.354839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.825397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.694215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bootstrap-10</th>\n",
       "      <th>ALL</th>\n",
       "      <td>94.560669</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.785441</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.358650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">tbprofiler</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">CAP</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">MGIT</th>\n",
       "      <th>bootstrap-7</th>\n",
       "      <th>ALL</th>\n",
       "      <td>76.250000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.823529</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.825397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bootstrap-8</th>\n",
       "      <th>ALL</th>\n",
       "      <td>70.068027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.866856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.261682</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bootstrap-9</th>\n",
       "      <th>ALL</th>\n",
       "      <td>67.816092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.466258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.934959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bootstrapped50</th>\n",
       "      <th>ALL</th>\n",
       "      <td>73.778591</td>\n",
       "      <td>0.858111</td>\n",
       "      <td>98.356579</td>\n",
       "      <td>0.165979</td>\n",
       "      <td>95.263516</td>\n",
       "      <td>0.46716</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entire</th>\n",
       "      <th>ALL</th>\n",
       "      <td>74.308300</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.434783</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.431472</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>188.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>566.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>828.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5408 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sensitivity  sensitivity_sem  \\\n",
       "set        drug method dataset        quality                                 \n",
       "basic      INH  UKMYC  bootstrap-0    ALL        88.803089              NaN   \n",
       "                                      HIGH       93.172691              NaN   \n",
       "                       bootstrap-1    ALL        91.828794              NaN   \n",
       "                                      HIGH       94.354839              NaN   \n",
       "                       bootstrap-10   ALL        94.560669              NaN   \n",
       "...                                                    ...              ...   \n",
       "tbprofiler CAP  MGIT   bootstrap-7    ALL        76.250000              NaN   \n",
       "                       bootstrap-8    ALL        70.068027              NaN   \n",
       "                       bootstrap-9    ALL        67.816092              NaN   \n",
       "                       bootstrapped50 ALL        73.778591         0.858111   \n",
       "                       entire         ALL        74.308300         0.000000   \n",
       "\n",
       "                                               specificity  specificity_sem  \\\n",
       "set        drug method dataset        quality                                 \n",
       "basic      INH  UKMYC  bootstrap-0    ALL        95.020747              NaN   \n",
       "                                      HIGH       96.414343              NaN   \n",
       "                       bootstrap-1    ALL        94.650206              NaN   \n",
       "                                      HIGH       96.825397              NaN   \n",
       "                       bootstrap-10   ALL        95.785441              NaN   \n",
       "...                                                    ...              ...   \n",
       "tbprofiler CAP  MGIT   bootstrap-7    ALL        98.823529              NaN   \n",
       "                       bootstrap-8    ALL        98.866856              NaN   \n",
       "                       bootstrap-9    ALL        98.466258              NaN   \n",
       "                       bootstrapped50 ALL        98.356579         0.165979   \n",
       "                       entire         ALL        98.434783         0.000000   \n",
       "\n",
       "                                                     PPV  PPV_sem     RR  \\\n",
       "set        drug method dataset        quality                              \n",
       "basic      INH  UKMYC  bootstrap-0    ALL      95.041322      NaN    NaN   \n",
       "                                      HIGH     96.265560      NaN    NaN   \n",
       "                       bootstrap-1    ALL      94.779116      NaN    NaN   \n",
       "                                      HIGH     96.694215      NaN    NaN   \n",
       "                       bootstrap-10   ALL      95.358650      NaN    NaN   \n",
       "...                                                  ...      ...    ...   \n",
       "tbprofiler CAP  MGIT   bootstrap-7    ALL      96.825397      NaN    NaN   \n",
       "                       bootstrap-8    ALL      96.261682      NaN    NaN   \n",
       "                       bootstrap-9    ALL      95.934959      NaN    NaN   \n",
       "                       bootstrapped50 ALL      95.263516  0.46716    NaN   \n",
       "                       entire         ALL      95.431472  0.00000  188.0   \n",
       "\n",
       "                                                 SR   UR   RS     SS   US  \\\n",
       "set        drug method dataset        quality                               \n",
       "basic      INH  UKMYC  bootstrap-0    ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "                                      HIGH      NaN  NaN  NaN    NaN  NaN   \n",
       "                       bootstrap-1    ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "                                      HIGH      NaN  NaN  NaN    NaN  NaN   \n",
       "                       bootstrap-10   ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "...                                             ...  ...  ...    ...  ...   \n",
       "tbprofiler CAP  MGIT   bootstrap-7    ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "                       bootstrap-8    ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "                       bootstrap-9    ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "                       bootstrapped50 ALL       NaN  NaN  NaN    NaN  NaN   \n",
       "                       entire         ALL      62.0  3.0  9.0  566.0  0.0   \n",
       "\n",
       "                                               Total  \n",
       "set        drug method dataset        quality         \n",
       "basic      INH  UKMYC  bootstrap-0    ALL        NaN  \n",
       "                                      HIGH       NaN  \n",
       "                       bootstrap-1    ALL        NaN  \n",
       "                                      HIGH       NaN  \n",
       "                       bootstrap-10   ALL        NaN  \n",
       "...                                              ...  \n",
       "tbprofiler CAP  MGIT   bootstrap-7    ALL        NaN  \n",
       "                       bootstrap-8    ALL        NaN  \n",
       "                       bootstrap-9    ALL        NaN  \n",
       "                       bootstrapped50 ALL        NaN  \n",
       "                       entire         ALL      828.0  \n",
       "\n",
       "[5408 rows x 13 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results={}\n",
    "PREDICTIONS.reset_index(inplace=True)\n",
    "PREDICTIONS.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "\n",
    "for i in tqdm(['basic','nulls','nulls+minors', \"tbprofiler\"]):\n",
    "    if i == 'tbprofiler':\n",
    "        predictions = pandas.read_csv(\"dat/tbprofiler_PREDICTIONS.csv\")\n",
    "        predictions.set_index(['ENA_RUN_ACCESSION', 'DRUG'], inplace=True)\n",
    "        df = copy.deepcopy(predictions.join(PHENOTYPES, how='inner'))\n",
    "    else:\n",
    "        predictions = PREDICTIONS[PREDICTIONS.SET==i]\n",
    "        df = copy.deepcopy(predictions.join(PHENOTYPES, how='inner'))\n",
    "\n",
    "    plate_df = df[df.PHENOTYPE_METHOD=='UKMYC']\n",
    "    plate_df.reset_index(inplace=True)\n",
    "    table = []\n",
    "    table = build_exact_table(table, plate_df, plate_drugs, ['HIGH', 'ALL'], i, 'UKMYC', include_fails=False)\n",
    "    table = build_bootstrapped_table(table,plate_df, plate_drugs, ['HIGH', 'ALL'], i, 'UKMYC', include_fails=False)\n",
    "\n",
    "    mgit_df = df[df.PHENOTYPE_METHOD=='MGIT']\n",
    "    mgit_df.reset_index(inplace=True)\n",
    "    table = build_exact_table(table, mgit_df, mgit_drugs, ['ALL'], i, 'MGIT', include_fails=False)\n",
    "    table = build_bootstrapped_table(table,mgit_df, mgit_drugs, ['ALL'], i, 'MGIT', include_fails=False)\n",
    "    results[i] = pandas.DataFrame(table, columns=['set','drug','method','quality','dataset','sensitivity','sensitivity_sem','specificity','specificity_sem','PPV','PPV_sem', 'RR', 'SR', 'UR', 'RS', 'SS', 'US', 'Total'])\n",
    "\n",
    "    results[i].set_index(['method','drug'],inplace=True)\n",
    "    results[i] = results[i][results[i].index.isin(list(results_drug_order.index))]\n",
    "    results[i].reset_index(inplace=True)\n",
    "    results[i].drug = results[i].drug.astype('category')\n",
    "    results[i].drug = results[i].drug.cat.set_categories(who_drugs)\n",
    "    results[i].sort_values(['drug', 'dataset', 'quality'], inplace=True)\n",
    "    results[i].set_index(['set','drug','method','dataset', 'quality'],inplace=True)\n",
    "\n",
    "results = pandas.concat(results.values())\n",
    "results.to_csv('dat/RESULTS.csv')\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the number of Fail results when we allow these rules to contribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 194 rows with a Fail call in the EFFECTS table (but these can be one variant affecting multiple drugs)\n",
      "Of the 194 Fails, 17 are duplicates\n",
      "And the most common genes are:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GENE\n",
       "rrs       106\n",
       "rpoB       20\n",
       "katG       17\n",
       "pncA       10\n",
       "embB        9\n",
       "Rv0678      4\n",
       "gid         2\n",
       "eis         2\n",
       "tlyA        2\n",
       "rplC        2\n",
       "ethA        1\n",
       "fabG1       1\n",
       "gyrB        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFFECTS.reset_index(inplace=True)\n",
    "EFFECTS['GENE_MUTATION'] = EFFECTS.GENE + '_' + EFFECTS.MUTATION\n",
    "print(f\"There are {len(EFFECTS[(EFFECTS.SET=='nulls') & (EFFECTS.PREDICTION=='F')])} rows with a Fail call in the EFFECTS table (but these can be one variant affecting multiple drugs)\")\n",
    "\n",
    "df = EFFECTS[(EFFECTS.SET=='nulls') & (EFFECTS.PREDICTION=='F')]\n",
    "df.set_index(['ENA_RUN_ACCESSION', 'GENE', 'MUTATION'], inplace=True)\n",
    "print(f\"Of the {len(df)} Fails, {df.index.duplicated().sum()} are duplicates\")\n",
    "\n",
    "df = df[~df.index.duplicated()]\n",
    "df.reset_index(inplace=True)\n",
    "print(\"And the most common genes are:\")\n",
    "df.GENE.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at which the RAVs most likely to be affected we can see how STM is affected more than AMI and KAN "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GENE_MUTATION\n",
       "rrs_c517x     48\n",
       "rrs_a514x     46\n",
       "embB_G406X     8\n",
       "katG_S315X     7\n",
       "rrs_g1484x     5\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.GENE_MUTATION.value_counts()[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>set</th>\n",
       "      <th>drug</th>\n",
       "      <th>method</th>\n",
       "      <th>dataset</th>\n",
       "      <th>quality</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>sensitivity_sem</th>\n",
       "      <th>specificity</th>\n",
       "      <th>specificity_sem</th>\n",
       "      <th>PPV</th>\n",
       "      <th>PPV_sem</th>\n",
       "      <th>RR</th>\n",
       "      <th>SR</th>\n",
       "      <th>UR</th>\n",
       "      <th>RS</th>\n",
       "      <th>SS</th>\n",
       "      <th>US</th>\n",
       "      <th>Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3535</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>12.820513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.476510</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.258065</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>548.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>674.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3223</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>LZD</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>42.857143</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.847095</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.428571</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>599.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>717.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3327</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>MXF</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>90.128755</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.360161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86.419753</td>\n",
       "      <td>0.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>730.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>DLM</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>22.222222</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>523.0</td>\n",
       "      <td>161.0</td>\n",
       "      <td>747.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3067</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>EMB</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>89.867841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>81.529851</td>\n",
       "      <td>0.0</td>\n",
       "      <td>67.326733</td>\n",
       "      <td>0.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>408.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>765.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>LEV</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>87.307692</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.584440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.653061</td>\n",
       "      <td>0.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>466.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>788.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3743</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>AMI</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>84.426230</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.307958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.095238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>206.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>537.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>823.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3899</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ETH</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>77.952756</td>\n",
       "      <td>0.0</td>\n",
       "      <td>85.478548</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69.230769</td>\n",
       "      <td>0.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>397.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>861.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2911</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>RIF</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>96.583144</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.671982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.711061</td>\n",
       "      <td>0.0</td>\n",
       "      <td>424.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>405.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>878.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4003</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>KAN</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>85.328185</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.546042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.086957</td>\n",
       "      <td>0.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>564.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>879.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>INH</td>\n",
       "      <td>UKMYC</td>\n",
       "      <td>entire</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>95.394737</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.495495</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.604396</td>\n",
       "      <td>0.0</td>\n",
       "      <td>435.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>399.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>904.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               set drug method dataset quality  sensitivity  sensitivity_sem  \\\n",
       "3535  nulls+minors  CFZ  UKMYC  entire    HIGH    12.820513              0.0   \n",
       "3223  nulls+minors  LZD  UKMYC  entire    HIGH    42.857143              0.0   \n",
       "3327  nulls+minors  MXF  UKMYC  entire    HIGH    90.128755              0.0   \n",
       "3639  nulls+minors  DLM  UKMYC  entire    HIGH    22.222222              0.0   \n",
       "3067  nulls+minors  EMB  UKMYC  entire    HIGH    89.867841              0.0   \n",
       "3431  nulls+minors  LEV  UKMYC  entire    HIGH    87.307692              0.0   \n",
       "3743  nulls+minors  AMI  UKMYC  entire    HIGH    84.426230              0.0   \n",
       "3899  nulls+minors  ETH  UKMYC  entire    HIGH    77.952756              0.0   \n",
       "2911  nulls+minors  RIF  UKMYC  entire    HIGH    96.583144              0.0   \n",
       "4003  nulls+minors  KAN  UKMYC  entire    HIGH    85.328185              0.0   \n",
       "2807  nulls+minors  INH  UKMYC  entire    HIGH    95.394737              0.0   \n",
       "\n",
       "      specificity  specificity_sem         PPV  PPV_sem     RR    SR    UR  \\\n",
       "3535    96.476510              0.0   32.258065      0.0   10.0  60.0   8.0   \n",
       "3223    99.847095              0.0   96.428571      0.0   27.0  33.0   3.0   \n",
       "3327    93.360161              0.0   86.419753      0.0  210.0  19.0   4.0   \n",
       "3639   100.000000              0.0  100.000000      0.0   14.0  33.0  16.0   \n",
       "3067    81.529851              0.0   67.326733      0.0  204.0  15.0   8.0   \n",
       "3431    96.584440              0.0   92.653061      0.0  227.0  26.0   7.0   \n",
       "3743    99.307958              0.0   98.095238      0.0  206.0  34.0   4.0   \n",
       "3899    85.478548              0.0   69.230769      0.0  198.0  20.0  36.0   \n",
       "2911    95.671982              0.0   95.711061      0.0  424.0  15.0   0.0   \n",
       "4003    98.546042              0.0   96.086957      0.0  221.0  32.0   6.0   \n",
       "2807    95.495495              0.0   95.604396      0.0  435.0  16.0   5.0   \n",
       "\n",
       "        RS     SS     US  Total  \n",
       "3535  21.0  548.0   27.0  674.0  \n",
       "3223   1.0  599.0   54.0  717.0  \n",
       "3327  33.0  426.0   38.0  730.0  \n",
       "3639   0.0  523.0  161.0  747.0  \n",
       "3067  99.0  408.0   29.0  765.0  \n",
       "3431  18.0  466.0   43.0  788.0  \n",
       "3743   4.0  537.0   37.0  823.0  \n",
       "3899  88.0  397.0  121.0  861.0  \n",
       "2911  19.0  405.0   15.0  878.0  \n",
       "4003   9.0  564.0   46.0  879.0  \n",
       "2807  20.0  399.0   25.0  904.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.reset_index(inplace=True)\n",
    "df = results[(results.set==\"nulls+minors\") & (results.method==\"UKMYC\") & (results.dataset==\"entire\") & (results.quality=='HIGH')]\n",
    "df.sort_values(by='Total')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:22: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row1.RR), int(row1.SR+row1.UR), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:23: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row2.RR), int(row2.SR+row2.UR), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:31: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row1.SS), int(row1.RS), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:32: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row2.SS), int(row2.RS), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:40: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row1.RR), int(row1.RS), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:41: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row2.RR), int(row2.RS), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:50: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row1.SS), int(row1.SR+row1.UR), ],)\n",
      "/var/folders/84/9nb2gsqj4hlcxztvwf19wg680000gq/T/ipykernel_90466/1814392013.py:51: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  tab.append([int(row2.SS), int(row2.SR+row2.UR), ],)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1148: RuntimeWarning: divide by zero encountered in log\n",
      "  lor = np.log(self.oddsratio_pooled)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1115: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  lor_va = np.sum(self._apd * self._ad / self._n**2) / adns**2\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1119: RuntimeWarning: divide by zero encountered in scalar divide\n",
      "  mid /= (adns * bcns)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1087: RuntimeWarning: divide by zero encountered in log\n",
      "  return np.log(self.oddsratio_pooled)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1219: RuntimeWarning: divide by zero encountered in divide\n",
      "  1 / (self._dma + e11))\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1222: RuntimeWarning: invalid value encountered in divide\n",
      "  statistic = np.sum((table[0, 0, :] - e11)**2 / v11)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1215: RuntimeWarning: invalid value encountered in divide\n",
      "  e11 = (-b + dr) / (2*a)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1077: RuntimeWarning: divide by zero encountered in scalar divide\n",
      "  odds_ratio = np.sum(self._ad / self._n) / np.sum(self._bc / self._n)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1121: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  lor_va += np.sum((1 - self._apd / self._n) *\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1214: RuntimeWarning: invalid value encountered in subtract\n",
      "  dr = np.sqrt(b**2 - 4*a*c)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1077: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  odds_ratio = np.sum(self._ad / self._n) / np.sum(self._bc / self._n)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1119: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  mid /= (adns * bcns)\n",
      "/Users/fowler/Library/Python/3.12/lib/python/site-packages/statsmodels/stats/contingency_tables.py:1058: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  statistic /= denom\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>INH</td>\n",
       "      <td>0.118</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.430</td>\n",
       "      <td>0.174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>RIF</td>\n",
       "      <td>0.472</td>\n",
       "      <td>0.525</td>\n",
       "      <td>0.354</td>\n",
       "      <td>0.705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>PZA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>EMB</td>\n",
       "      <td>0.388</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.544</td>\n",
       "      <td>0.679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>BDQ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>AMI</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.783</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>STM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>ETH</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.706</td>\n",
       "      <td>0.783</td>\n",
       "      <td>0.387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.765</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CAP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "0          basic      ALL         basic     HIGH  INH        0.118   \n",
       "1          basic      ALL         basic     HIGH  RIF        0.472   \n",
       "2          basic      ALL         basic     HIGH  PZA          NaN   \n",
       "3          basic      ALL         basic     HIGH  EMB        0.388   \n",
       "4          basic      ALL         basic     HIGH  BDQ          NaN   \n",
       "..           ...      ...           ...      ...  ...          ...   \n",
       "66  nulls+minors      ALL  nulls+minors     HIGH  AMI        0.005   \n",
       "67  nulls+minors      ALL  nulls+minors     HIGH  STM          NaN   \n",
       "68  nulls+minors      ALL  nulls+minors     HIGH  ETH        0.175   \n",
       "69  nulls+minors      ALL  nulls+minors     HIGH  KAN        0.014   \n",
       "70  nulls+minors      ALL  nulls+minors     HIGH  CAP          NaN   \n",
       "\n",
       "    specificity    PPV    NPV  \n",
       "0         0.522  0.430  0.174  \n",
       "1         0.525  0.354  0.705  \n",
       "2           NaN    NaN    NaN  \n",
       "3         0.148  0.544  0.679  \n",
       "4           NaN    NaN    NaN  \n",
       "..          ...    ...    ...  \n",
       "66        0.783  0.957  0.027  \n",
       "67          NaN    NaN    NaN  \n",
       "68        0.706  0.783  0.387  \n",
       "69        0.765  0.993  0.051  \n",
       "70          NaN    NaN    NaN  \n",
       "\n",
       "[71 rows x 9 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conditions = [  (('basic', 'ALL'), ('basic', 'HIGH')),\n",
    "                (('basic', 'ALL'), ('nulls+minors', 'ALL')),\n",
    "                (('basic', 'ALL'), ('nulls+minors', 'HIGH')),\n",
    "                (('basic', 'HIGH'), ('nulls+minors', 'HIGH')),\n",
    "                (('nulls+minors', 'ALL'), ('nulls+minors', 'HIGH')),\n",
    "            ]\n",
    "\n",
    "df = results[results.dataset=='entire']\n",
    "\n",
    "statistical_rows = []\n",
    "\n",
    "for i in conditions:\n",
    "    a = df[(df.set==i[0][0]) & (df.quality==i[0][1])]\n",
    "    b = df[(df.set==i[1][0]) & (df.quality==i[1][1])]\n",
    "    for drug in a.drug.unique():\n",
    "        \n",
    "        row1 = a[a.drug==drug]\n",
    "        row2 = b[b.drug==drug]\n",
    "\n",
    "        if not row1.empty and not row2.empty:\n",
    "            tab = []\n",
    "            tab.append([int(row1.RR), int(row1.SR+row1.UR), ],)\n",
    "            tab.append([int(row2.RR), int(row2.SR+row2.UR), ],)        \n",
    "            st = sm.stats.StratifiedTable([tab])\n",
    "            st_results=st.summary().data\n",
    "            for st_line in st_results:\n",
    "                if st_line[0]=='Test of OR=1':\n",
    "                    sens_pvalue=float(st_line[2])\n",
    "            \n",
    "            tab = []\n",
    "            tab.append([int(row1.SS), int(row1.RS), ],)\n",
    "            tab.append([int(row2.SS), int(row2.RS), ],)        \n",
    "            st = sm.stats.StratifiedTable([tab])\n",
    "            st_results=st.summary().data\n",
    "            for st_line in st_results:\n",
    "                if st_line[0]=='Test of OR=1':\n",
    "                    spec_pvalue=float(st_line[2])\n",
    "\n",
    "            tab = []\n",
    "            tab.append([int(row1.RR), int(row1.RS), ],)\n",
    "            tab.append([int(row2.RR), int(row2.RS), ],)        \n",
    "            st = sm.stats.StratifiedTable([tab])\n",
    "            st_results=st.summary().data\n",
    "            for st_line in st_results:\n",
    "                if st_line[0]=='Test of OR=1':\n",
    "                    ppv_pvalue=float(st_line[2])\n",
    "\n",
    "\n",
    "            tab = []\n",
    "            tab.append([int(row1.SS), int(row1.SR+row1.UR), ],)\n",
    "            tab.append([int(row2.SS), int(row2.SR+row2.UR), ],)        \n",
    "            st = sm.stats.StratifiedTable([tab])\n",
    "            st_results=st.summary().data\n",
    "            for st_line in st_results:\n",
    "                if st_line[0]=='Test of OR=1':\n",
    "                    npv_pvalue=float(st_line[2])\n",
    "\n",
    "            statistical_rows.append([i[0][0], i[0][1], i[1][0], i[1][1], drug, sens_pvalue, spec_pvalue, ppv_pvalue, npv_pvalue])\n",
    "        else:\n",
    "            statistical_rows.append([i[0][0], i[0][1], i[1][0], i[1][1], drug, None, None, None, None])\n",
    "\n",
    "statistical_table = pandas.DataFrame(statistical_rows, columns=['dataset1', 'quality1', 'dataset2', 'quality2', 'drug', 'sensitivity', 'specificity', 'PPV', 'NPV'])\n",
    "statistical_table.to_csv('dat/RESULTS_STATISTICAL_SIGNIFICANCE.csv')\n",
    "statistical_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.793</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.334</td>\n",
       "      <td>0.613</td>\n",
       "      <td>0.689</td>\n",
       "      <td>0.441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.430</td>\n",
       "      <td>0.683</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.238</td>\n",
       "      <td>0.612</td>\n",
       "      <td>0.689</td>\n",
       "      <td>0.310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.765</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "13         basic      ALL         basic     HIGH  KAN        0.029   \n",
       "28         basic      ALL  nulls+minors      ALL  KAN        0.334   \n",
       "43         basic      ALL  nulls+minors     HIGH  KAN        0.001   \n",
       "55         basic     HIGH  nulls+minors     HIGH  KAN        0.238   \n",
       "69  nulls+minors      ALL  nulls+minors     HIGH  KAN        0.014   \n",
       "\n",
       "    specificity    PPV    NPV  \n",
       "13        0.793  0.993  0.098  \n",
       "28        0.613  0.689  0.441  \n",
       "43        0.430  0.683  0.008  \n",
       "55        0.612  0.689  0.310  \n",
       "69        0.765  0.993  0.051  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[statistical_table.drug=='KAN']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>AMI</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.812</td>\n",
       "      <td>0.966</td>\n",
       "      <td>0.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.793</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LZD</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.745</td>\n",
       "      <td>0.694</td>\n",
       "      <td>0.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.765</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "10         basic      ALL         basic     HIGH  AMI        0.010   \n",
       "13         basic      ALL         basic     HIGH  KAN        0.029   \n",
       "61  nulls+minors      ALL  nulls+minors     HIGH  LZD        0.047   \n",
       "69  nulls+minors      ALL  nulls+minors     HIGH  KAN        0.014   \n",
       "\n",
       "    specificity    PPV    NPV  \n",
       "10        0.812  0.966  0.050  \n",
       "13        0.793  0.993  0.098  \n",
       "61        0.745  0.694  0.007  \n",
       "69        0.765  0.993  0.051  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.sensitivity<0.05) & (statistical_table.sensitivity>=0.01)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.864</td>\n",
       "      <td>0.558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>CAP</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.722</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>INH</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.954</td>\n",
       "      <td>0.868</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LZD</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.833</td>\n",
       "      <td>0.979</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.383</td>\n",
       "      <td>0.885</td>\n",
       "      <td>0.051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.759</td>\n",
       "      <td>0.863</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.430</td>\n",
       "      <td>0.683</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>AMI</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.783</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "23         basic      ALL  nulls+minors      ALL  CFZ        0.006   \n",
       "29         basic      ALL  nulls+minors      ALL  CAP        0.002   \n",
       "30         basic      ALL  nulls+minors     HIGH  INH        0.007   \n",
       "35         basic      ALL  nulls+minors     HIGH  LZD        0.004   \n",
       "36         basic      ALL  nulls+minors     HIGH  MXF        0.004   \n",
       "37         basic      ALL  nulls+minors     HIGH  LEV        0.003   \n",
       "43         basic      ALL  nulls+minors     HIGH  KAN        0.001   \n",
       "66  nulls+minors      ALL  nulls+minors     HIGH  AMI        0.005   \n",
       "\n",
       "    specificity    PPV    NPV  \n",
       "23        0.055  0.864  0.558  \n",
       "29        0.990  0.722  0.015  \n",
       "30        0.954  0.868  0.016  \n",
       "35        0.833  0.979  0.002  \n",
       "36        0.383  0.885  0.051  \n",
       "37        0.759  0.863  0.016  \n",
       "43        0.430  0.683  0.008  \n",
       "66        0.783  0.957  0.027  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.sensitivity<0.01) & (statistical_table.sensitivity>=0.001)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>BDQ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.434</td>\n",
       "      <td>0.850</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>AMI</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dataset1 quality1      dataset2 quality2 drug  sensitivity  specificity  \\\n",
       "19    basic      ALL  nulls+minors      ALL  BDQ          0.0        0.434   \n",
       "40    basic      ALL  nulls+minors     HIGH  AMI          0.0        0.522   \n",
       "\n",
       "      PPV    NPV  \n",
       "19  0.850  0.000  \n",
       "40  0.718  0.004  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.sensitivity<0.001) & (statistical_table.sensitivity>=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset1, quality1, dataset2, quality2, drug, sensitivity, specificity, PPV, NPV]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.specificity<0.05)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LZD</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.389</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>CAP</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.722</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>INH</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.954</td>\n",
       "      <td>0.868</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.759</td>\n",
       "      <td>0.863</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>AMI</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.783</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "5          basic      ALL         basic     HIGH  LZD        0.067   \n",
       "29         basic      ALL  nulls+minors      ALL  CAP        0.002   \n",
       "30         basic      ALL  nulls+minors     HIGH  INH        0.007   \n",
       "37         basic      ALL  nulls+minors     HIGH  LEV        0.003   \n",
       "66  nulls+minors      ALL  nulls+minors     HIGH  AMI        0.005   \n",
       "\n",
       "    specificity    PPV    NPV  \n",
       "5         0.389  0.367  0.010  \n",
       "29        0.990  0.722  0.015  \n",
       "30        0.954  0.868  0.016  \n",
       "37        0.759  0.863  0.016  \n",
       "66        0.783  0.957  0.027  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.NPV<0.05) & (statistical_table.NPV>=0.01)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LZD</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.833</td>\n",
       "      <td>0.979</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>AMI</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>KAN</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.430</td>\n",
       "      <td>0.683</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>LZD</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.745</td>\n",
       "      <td>0.694</td>\n",
       "      <td>0.007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "35         basic      ALL  nulls+minors     HIGH  LZD        0.004   \n",
       "40         basic      ALL  nulls+minors     HIGH  AMI        0.000   \n",
       "43         basic      ALL  nulls+minors     HIGH  KAN        0.001   \n",
       "61  nulls+minors      ALL  nulls+minors     HIGH  LZD        0.047   \n",
       "\n",
       "    specificity    PPV    NPV  \n",
       "35        0.833  0.979  0.002  \n",
       "40        0.522  0.718  0.004  \n",
       "43        0.430  0.683  0.008  \n",
       "61        0.745  0.694  0.007  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.NPV<0.01) & (statistical_table.NPV>=0.001)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.566</td>\n",
       "      <td>0.950</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>DLM</td>\n",
       "      <td>0.116</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.395</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>BDQ</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.434</td>\n",
       "      <td>0.850</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>DLM</td>\n",
       "      <td>0.065</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.378</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.902</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>DLM</td>\n",
       "      <td>0.091</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.391</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "8          basic      ALL         basic     HIGH  CFZ        0.566   \n",
       "9          basic      ALL         basic     HIGH  DLM        0.116   \n",
       "19         basic      ALL  nulls+minors      ALL  BDQ        0.000   \n",
       "38         basic      ALL  nulls+minors     HIGH  CFZ        0.094   \n",
       "39         basic      ALL  nulls+minors     HIGH  DLM        0.065   \n",
       "64  nulls+minors      ALL  nulls+minors     HIGH  CFZ        0.808   \n",
       "65  nulls+minors      ALL  nulls+minors     HIGH  DLM        0.091   \n",
       "\n",
       "    specificity    PPV  NPV  \n",
       "8         0.950  0.027  0.0  \n",
       "9         0.372  0.395  0.0  \n",
       "19        0.434  0.850  0.0  \n",
       "38        0.083  0.033  0.0  \n",
       "39        0.372  0.378  0.0  \n",
       "64        0.902  0.010  0.0  \n",
       "65        0.372  0.391  0.0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.NPV<0.001) & (statistical_table.NPV>=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>basic</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.566</td>\n",
       "      <td>0.950</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>basic</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>ALL</td>\n",
       "      <td>nulls+minors</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>CFZ</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.902</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset1 quality1      dataset2 quality2 drug  sensitivity  \\\n",
       "8          basic      ALL         basic     HIGH  CFZ        0.566   \n",
       "38         basic      ALL  nulls+minors     HIGH  CFZ        0.094   \n",
       "64  nulls+minors      ALL  nulls+minors     HIGH  CFZ        0.808   \n",
       "\n",
       "    specificity    PPV  NPV  \n",
       "8         0.950  0.027  0.0  \n",
       "38        0.083  0.033  0.0  \n",
       "64        0.902  0.010  0.0  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.PPV<0.05) & (statistical_table.PPV>=0.01)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset1, quality1, dataset2, quality2, drug, sensitivity, specificity, PPV, NPV]\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.PPV<0.01) & (statistical_table.PPV>=0.001)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset1</th>\n",
       "      <th>quality1</th>\n",
       "      <th>dataset2</th>\n",
       "      <th>quality2</th>\n",
       "      <th>drug</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset1, quality1, dataset2, quality2, drug, sensitivity, specificity, PPV, NPV]\n",
       "Index: []"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistical_table[(statistical_table.PPV<0.001) & (statistical_table.PPV>=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
